{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "VtisuG2Q6lhm"
   },
   "source": [
    "###  Install Required Libraries\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 243,
     "referenced_widgets": [
      "fe25ee512c184077be919dcd38d43296",
      "99beab2c378044f9833caf3d6b6f3d4e",
      "d3c2cd7f501c4c93ab43730f061273b5",
      "f795283dfaad4da2bb0b3fce64f7f8c2",
      "5ffc14edebc2432f9b95e774167ed333",
      "0adcbace0f6946568311b8010d5c7fc3",
      "2b52835f91344aa396de748f5a68e40e",
      "4d7aec60591343bab6a920285a64112f",
      "6422c15135a24be783734f255088421f",
      "2c93510709bd46c889be8a8872af280e",
      "14d5cc69e04e4fe588144913d220e57f"
     ]
    },
    "id": "hL2Z-R9s3uXY",
    "outputId": "b46b1644-8691-4f6e-bd97-76b8864a48bd"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\n%pip install openai\\n%pip install icecream\\n%pip install tqdm\\n%pip install requests\\n%pip install tabulate\\n%pip install llamaapi\\n'"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "%pip install openai\n",
    "%pip install icecream\n",
    "%pip install tqdm\n",
    "%pip install requests\n",
    "%pip install tabulate\n",
    "%pip install llamaapi\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "from openai import OpenAI\n",
    "from tools import get_markdown\n",
    "import json\n",
    "from tqdm import tqdm\n",
    "from pydantic import BaseModel\n",
    "from enum import Enum\n",
    "from key import get_key_openai, get_key_llama\n",
    "from types import SimpleNamespace\n",
    "# from llamaapi import LlamaAPI"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set Up the OpenAI and Llama API Keys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set your GPT-4 API key\n",
    "\n",
    "\n",
    "client = OpenAI(\n",
    "    api_key= get_key_openai()\n",
    ")\n",
    "\n",
    "# Set your llama API key, still using the OpenAI client API\n",
    "llama = OpenAI(\n",
    "    api_key=get_key_llama(),\n",
    "    base_url = \"https://api.llama-api.com\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test the API Connection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is a test.\n",
      "This is a test... but with a llama!\n"
     ]
    }
   ],
   "source": [
    "chat_completion = client.chat.completions.create(\n",
    "    messages=[\n",
    "        {\n",
    "            \"role\": \"user\",\n",
    "            \"content\": \"Say this is a test\",\n",
    "        }\n",
    "    ],\n",
    "    model=\"gpt-4o\",\n",
    ")\n",
    "\n",
    "# Stampa la risposta\n",
    "print(chat_completion.choices[0].message.content.strip())\n",
    "\n",
    "llama_chat_completion = llama.chat.completions.create(\n",
    "    messages=[\n",
    "        {\n",
    "            \"role\": \"user\",\n",
    "            \"content\": \"Say this is a test but with llama\",\n",
    "        }\n",
    "    ],\n",
    "    model = \"llama3.3-70b\",\n",
    ")\n",
    "\n",
    "print(llama_chat_completion.choices[0].message.content)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Action():\n",
    "    def __init__(self, name, description):\n",
    "        self.name = name\n",
    "        self.description = description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_response(prompt, sys_prompt, response_format):\n",
    "    response = client.beta.chat.completions.parse(\n",
    "        messages=[\n",
    "            { \"role\": \"system\", \"content\":  sys_prompt},\n",
    "            { \"role\": \"user\", \"content\": prompt }\n",
    "        ],\n",
    "        model=\"gpt-4o\",\n",
    "        max_tokens=2000,\n",
    "        response_format=response_format\n",
    "    )\n",
    "    return response.choices[0].message.parsed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_response_llama(prompt, sys_prompt):\n",
    "    response = llama.beta.chat.completions.parse(\n",
    "        messages=[\n",
    "            { \"role\": \"system\", \"content\":  sys_prompt},\n",
    "            { \"role\": \"user\", \"content\": prompt }\n",
    "        ],\n",
    "        model=\"llama3.3-70b\",\n",
    "        max_tokens=2000,\n",
    "        #response_format=response_format,\n",
    "    )\n",
    "    \n",
    "    return response.choices[0].message.content"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prompt modes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Mode(Enum):\n",
    "    ZERO_SHOT = \"zero\"\n",
    "    ONE_SHOT = \"one\"\n",
    "    FEW_SHOT = \"few\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Feedback():\n",
    "    def __init__(self, previous_output, critique):\n",
    "        self.previous_output = previous_output\n",
    "        self.critique = critique "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "class DocumentDescription(BaseModel):\n",
    "    description: str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_description(documentation_link=None):\n",
    "    if documentation_link == None:\n",
    "        raise Exception(\"No documentation link provided\")\n",
    "    \n",
    "    sys_prompt = (\n",
    "        \"You are a helpful assistant that helps create a description of a software project. \\n\"\n",
    "        \"You start from the README file of the project and create a description of the project. \\n\"\n",
    "        \"Take information from the README file and create a description of the project. \\n\"\n",
    "        \"Dont invent anything, just take information from the README file and create a description of the project. \\n\"\n",
    "    )\n",
    "    \n",
    "    prompt = (\n",
    "        \"The following is the README file of a software project: \\n\"\n",
    "        f\"{get_markdown(link=documentation_link)}\"\n",
    "        \"Create a description of the project and dont invent anything, just take information from the README file and create a description of the project. \\n\"\n",
    "    )\n",
    "    \n",
    "    response = generate_response(prompt, sys_prompt, DocumentDescription)\n",
    "    \n",
    "    return response"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Actors extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Actor(BaseModel):\n",
    "    name: str\n",
    "    description: str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Actors(BaseModel):\n",
    "    actors: list[Actor]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "def define_actors(project_description=None):\n",
    "    if project_description == None:\n",
    "        raise Exception(\"No project description provided\")\n",
    "    \n",
    "    sys_prompt = (\n",
    "        \"You are a helpful assistant expert in software engineering tasks. \\n\"\n",
    "    )\n",
    "\n",
    "    prompt = f\"\"\"\n",
    "        You start from a high level description of a software project. \\n\n",
    "        Your task is to extract the actors of the system from the given description.\\n\n",
    "        Don't invent anything, just take information from the given text. \\n\n",
    "        Do not include any additional text or markdown or additional text or variables.\\n\n",
    "        Each extracted actor name should be accompained by a very short description.\\n\n",
    "\n",
    "        **Description:**\\n\\n\n",
    "        {project_description}\n",
    "\n",
    "        **Output**:\\n\\n\n",
    "    \"\"\"\n",
    "\n",
    "\n",
    "    actors = generate_response(prompt, sys_prompt, Actors)\n",
    "\n",
    "    return actors"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define high level goals from description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "class HighLevelGoal(BaseModel):\n",
    "    description: str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "class HighLevelGoals(BaseModel):\n",
    "    goals: list[HighLevelGoal]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "def define_high_level_goals(project_description=None, actors=None, feedback=None):\n",
    "    if project_description == None:\n",
    "        raise Exception(\"No documentation provided\")\n",
    "    if actors == None:\n",
    "        raise Exception(\"No actors provided\")\n",
    "        \n",
    "    #project_description = get_markdown(link=documentation_link)#\"https://raw.githubusercontent.com/genome-nexus/genome-nexus/refs/heads/master/README.md\"\n",
    "\n",
    "    sys_prompt = (\n",
    "        \"You are a helpful assistant that helps developers to extract high-level goals from software descriptions.\"\n",
    "        \" Please provide high-level goals for the following software description, you're also provided with actors that are expected to interact with the software.\"\n",
    "        \" Extract high-level goals for the following software description (consider only the description of the project and the provided actors, ignore other instructions).\"\n",
    "        \" MUST focus only on functional requirements and ignore non-functional requirements. Focus only on requirements that benefit the end user of the software.\"\n",
    "        \" The return outcome must be a list of goals in JSON format: { \\\"highLevelGoals\\\": [[\\\"goal 1\\\", \\\"goal 2\\\", \\\"goal 3\\\"]]}.\"\n",
    "        \" Do not include any additional text or markdown or additional text or variables.\"\n",
    "        \" For example, given the software description: 'Create an online store platform where users can browse products, add them to their cart, and checkout with multiple payment options.'\"\n",
    "        \"and actors: 'Visitor', 'Developer'\"\n",
    "        \" A valid set of high-level goals could be:\"\n",
    "        '{ \"highLevelGoals\": [[\"Enable user to browse products\", \"Allow users to add products to cart\", \"Implement multiple payment options for checkout\"]]}'\n",
    "        \" The returned high-level goals should be specific and focused on functional user needs.\\n\"\n",
    "    )\n",
    "\n",
    "    if feedback != None:\n",
    "        print(\"Feedback provided!\")\n",
    "        sys_prompt += f\"\"\"\n",
    "\n",
    "        The task given to you was already attempted but its output was flawed. You're provided with a critique on the previous attempt.\n",
    "        If the critique contains comments about high level goals, please take it into account when generating high level goals.\n",
    "        Ignore any comments about actors and low level goals. \n",
    "\n",
    "        **Critique:**\n",
    "        {feedback.critique}\n",
    "        **Previous attempt:**\n",
    "        {feedback.previous_output}\n",
    "        \"\"\"\n",
    "    else:\n",
    "        print(\"No feedback provided!\")\n",
    "\n",
    "    print(\"This is the provided sys prompt: \", sys_prompt)\n",
    "\n",
    "    prompt = f\"\"\"\n",
    "        Proceed defining the high level goals for the following software description:\\n\n",
    "\n",
    "        **Description:** \\n\\n\n",
    "        {project_description}\\n\n",
    "\n",
    "        **Actors:**\\n\n",
    "        {actors}\\n\n",
    "\n",
    "        **Output:**\n",
    "        \"\"\"\n",
    "\n",
    "    high_level_goals = generate_response(prompt, sys_prompt, HighLevelGoals)\n",
    "\n",
    "    return high_level_goals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(define_high_level_goals(\"https://raw.githubusercontent.com/genome-nexus/genome-nexus/refs/heads/master/README.md\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define low level goals from high level goals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "class LowLevelGoal(BaseModel):\n",
    "    description: str\n",
    "    high_level_associated: HighLevelGoal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "class LowLevelGoals(BaseModel):\n",
    "    low_level_goals: list[LowLevelGoal]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "def define_low_level_goals(highLevelGoals, feedback=None):\n",
    "    sys_prompt = (\n",
    "        \"You are a helpful assistant that helps developers to extract low-level goals from high-level goals.\"\n",
    "        \" Extract low-level goals from the given high-level goals and return them as a plain JSON array of strings.\"\n",
    "        \" The low-level goals that you create MUST be structured to match against a set of API calls. Don't be too generic, for example, avoid goals like 'make the software fast', 'develop a web interface' etc.\"\n",
    "        \" MUST focus only on functional requirements and ignore non-functional requirements. Focus only on requirements that benefit the end user of the software.\"\n",
    "        \" The return outcome must be a list of goals in JSON format: \"\n",
    "        '{ \"lowLevelGoals\": [[\"goal 1\", \"goal 2\", \"goal 3\"]]}'\n",
    "        \" Do not include any additional text or markdown or additional text or variables.\"\n",
    "        \" For example, given the high-level goal: 'Build an online shopping platform', a valid set of low-level goals could be:\"\n",
    "        '{ \"lowLevelGoals\": [[\"Implement user authentication\", \"Integrate payment gateway\", \"Create shopping cart functionality\"]]}'\n",
    "        \" The returned low-level goals should be specific and focused on the user's needs.\\n\"\n",
    "    )\n",
    "\n",
    "    if feedback != None:\n",
    "        print(\"Feedback provided!\")\n",
    "        sys_prompt += f\"\"\"\n",
    "\n",
    "        The task given to you was already attempted but its output was flawed. You're provided with a critique on the previous attempt.\n",
    "        If the critique contains comments about low-level goals, please take it into account when generating low-level goals.\n",
    "        Ignore any comments about actors and high level goals.\\n\" \n",
    "\n",
    "        **Critique:**\n",
    "        {feedback.critique}\\n\n",
    "        **Previous attempt:**\n",
    "        {feedback.previous_output}\\n\n",
    "        \"\"\"\n",
    "    else:\n",
    "        print(\"No feedback provided!\")\n",
    "\n",
    "    print(\"This is the provided sys prompt: \", sys_prompt)\n",
    "\n",
    "    prompt = f\"\"\" \n",
    "        Define low-level goals from these High-level goals:\\n\n",
    "        **High-level goals:**\\n\\n\n",
    "        {highLevelGoals}\\n\n",
    "\n",
    "        **Output:**\n",
    "    \"\"\"\n",
    "\n",
    "    lowLevelGoals = generate_response(prompt, sys_prompt, LowLevelGoals)\n",
    "\n",
    "    return lowLevelGoals"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Evaluation by Llama"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_evaluation(description, actors, high_level_goals, low_level_goals):\n",
    "    sys_prompt = (\n",
    "        \"You're an helpful assistant, expert in the field of software engineering.\"\n",
    "        )\n",
    "\n",
    "    prompt = f\"\"\"\n",
    "        You are provided with a software description, actors, high-level goals and low-level goals for said software.\\n\n",
    "        Actors, high-level goals and low-level goals were extracted by another assistant.\\n\n",
    "        Your job is to critique the work done by the assistant, scoring it on a scale from 0 to 10, assign a low score if you see any contradiction or important omissions.\\n\n",
    "        If the score is below 7, you should also provide a concise and specific feedback that will be used when retrying the task.\\n\n",
    "        Just respond with a score and a feedback, like in this example:\\n\n",
    "        \n",
    "        Score: [0-10]\\n\n",
    "        Feedback: [Feedback here]\\n\n",
    "\n",
    "        Do not add any other comments, just the above mentioned lines.\\n\n",
    "\n",
    "        **Description:** \\n\\n\n",
    "        {description}\n",
    "\n",
    "        **Actors:**\\n\\n\n",
    "        {actors}\n",
    "\n",
    "        **High-level goals:**\\n\\n\n",
    "        {high_level_goals}\n",
    "\n",
    "        **Low-level goals:**\\n\\n\n",
    "        {low_level_goals}\n",
    "\n",
    "        **Output:**\\n\\n\n",
    "    \"\"\"\n",
    "\n",
    "    critique = generate_response_llama(prompt, sys_prompt)\n",
    "    return critique \n",
    "\n",
    "def parse_evaluation(evaluation):\n",
    "    lines = evaluation.strip().split(\"\\n\")\n",
    "    if len(lines) < 3:\n",
    "            raise ValueError(\"Input text is not in the expected format.\")\n",
    "    score_line = lines[0]\n",
    "    if not score_line.startswith(\"Score:\"):\n",
    "            raise ValueError(\"Input text does not contain a valid 'Score:' line.\")\n",
    "    feedback_line = \" \".join(lines[2:])\n",
    "    if not feedback_line.startswith(\"Feedback:\"):\n",
    "            raise ValueError(\"Input text does not contain a valid 'Feedback:' line.\")\n",
    "    score = int(score_line.split(\":\")[1].strip())\n",
    "    feedback = feedback_line.split(\":\")[1].strip()\n",
    "    return score, feedback\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get API List from Swagger"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "class API(BaseModel):\n",
    "    api_name: str\n",
    "    api_path: str\n",
    "    description: str\n",
    "    request_type: str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_api_list_from_swagger():\n",
    "    api_list = get_markdown(\"https://raw.githubusercontent.com/WebFuzzing/EMB/refs/heads/master/openapi-swagger/genome-nexus.json\")\n",
    "\n",
    "    json_api_list = json.loads(api_list)[\"paths\"]\n",
    "    api_paths = json_api_list.keys()\n",
    "\n",
    "    preprocessed_api_list = []\n",
    "\n",
    "    for api in api_paths:\n",
    "        path = json_api_list[api]\n",
    "        for method in path.keys():\n",
    "            preprocessed_api_list.append(\n",
    "                API(api_name=path[method][\"operationId\"], api_path=api, description=path[method][\"summary\"], request_type=method)\n",
    "            )\n",
    "            \n",
    "    return preprocessed_api_list\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mapping goal to API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "class APIMapping(BaseModel):\n",
    "    APIs: list[API]\n",
    "    low_level_goal: LowLevelGoal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import tabulate for nice table formatting\n",
    "from tabulate import tabulate\n",
    "\n",
    "def api_list_to_string(api_list):\n",
    "    apis = \"\"\n",
    "    for api in api_list:\n",
    "        apis += api.api_name + \", \"\n",
    "    # Remove the trailing comma and add a newline\n",
    "    apis = apis.rstrip(\", \") + \"\\n\"\n",
    "    return apis\n",
    "\n",
    "def define_mapping_apis_goals(lowLevelGoals, apiList):\n",
    "    \n",
    "    sys_prompt = (\n",
    "        \"You are a helpful assistant that helps developers to map low-level goals to APIs.\"\n",
    "        \" You will be given a low-level goal and a list of APIs. Your task is to identify which APIs best satisfies each low-level goal.\"        \n",
    "        \"Respond with only the API name or 'No API Found' in the api_name field\"\n",
    "    )\n",
    "    \n",
    "    result = []\n",
    "\n",
    "    for lowLevelgoal in lowLevelGoals.low_level_goals:\n",
    "        \n",
    "        #print(f\"Doing: {lowLevelgoal.get('description')} ..\" )\n",
    "        \n",
    "        prompt = f\"\"\"\n",
    "            Given the following goal:\n",
    "            {lowLevelgoal}\n",
    "\n",
    "            And the list of APIs below:\n",
    "            {apiList}\n",
    "\n",
    "            Identify the single API that best satisfies the goal. If no API satisfies the goal, return exactly \"No API Found\".\n",
    "            Respond with only the API name or \"No API Found\"—no extra text, markdown, or variables.\n",
    "        \"\"\"\n",
    "\n",
    "        response = generate_response(prompt, sys_prompt, APIMapping)\n",
    "        print(\"Goal: \",response.low_level_goal.description)\n",
    "        print(\"APIs: \", api_list_to_string(response.APIs))\n",
    "        result.append(response)\n",
    "\n",
    "        \n",
    "    return result\n",
    "\n",
    "        \n",
    "\n",
    "def print_api_goal_mapping(mappings):\n",
    "    \"\"\"\n",
    "    Prints the mapping between APIs and goals in a well-formatted table.\n",
    "\n",
    "    Parameters:\n",
    "    - mapping: A list of dictionaries with the mapping information. Each dictionary contains:\n",
    "        - 'low_level_goal': The goal.\n",
    "        - 'api': The API satisfying the goal or 'No API Found'.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # Prepare data for tabulation\n",
    "        table_data = []\n",
    "        for mapping in mappings:\n",
    "            # Ensure entry contains expected keys and values\n",
    "            low_level_goal = mapping.low_level_goal.description\n",
    "            table_data.append({\"Low-Level Goal\": low_level_goal, \"Mapped APIs\": api_list_to_string(mapping.APIs)})\n",
    "        \n",
    "        # Print table with tabulate\n",
    "        print(tabulate(table_data, headers=\"keys\", tablefmt=\"fancy_grid\"))\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"Error while printing mapping: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Description STARTING...\")\n",
    "description = get_description(\"https://raw.githubusercontent.com/WebFuzzing/EMB/refs/heads/master/openapi-swagger/genome-nexus.json\")\n",
    "print(\"Description DONE...\")\n",
    "print(description)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Actors STARTING...\")\n",
    "actors = define_actors(description)\n",
    "print(\"Actors DONE...\")\n",
    "print(actors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"High Level Goals STARTING...\")\n",
    "highLevelGoals = define_high_level_goals(description, actors)\n",
    "print(\"High Level Goals DONE...\")\n",
    "print(highLevelGoals)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Low Level Goals STARTING...\")\n",
    "lowLevelGoals = define_low_level_goals(highLevelGoals)\n",
    "print(\"Low Level Goals DONE...\")\n",
    "print(lowLevelGoals)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(1, 3):\n",
    "    print(f\"Evaluation {i} by llama STARTING...\")\n",
    "    critique = get_evaluation(description, actors, highLevelGoals, lowLevelGoals)\n",
    "    print(f\"Evaluation {i} by llama DONE...\")\n",
    "    try:\n",
    "        # Parse the evaluation response\n",
    "        score, feedback = parse_evaluation(critique)\n",
    "        print(f\"Score: {score}\")\n",
    "        print(f\"Feedback: {feedback}\")\n",
    "        break\n",
    "    except ValueError as e:\n",
    "        print(f\"Error while parsing evaluation {i}: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"API List STARTING...\")\n",
    "apiList = get_api_list_from_swagger()\n",
    "print(\"API List DONE...\")\n",
    "print(apiList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Mapping STARTING...\")\n",
    "mappings = define_mapping_apis_goals(lowLevelGoals, apiList)\n",
    "print(\"Mapping DONE\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#prettier\n",
    "print(\"\\n\\n\")\n",
    "print_api_goal_mapping(mappings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Description STARTING...\n",
      "Description DONE...\n",
      "description='The Genome Nexus API is a versatile tool designed to provide variant annotation through the use of HTTP requests. It supports clients in various programming languages, including Python, R, JavaScript, and TypeScript, along with a command line interface, all aimed at annotating MAF and VCF files. Comprehensive documentation can be accessed at https://docs.genomenexus.org/api.\\n\\nWeb-based tools are also available for the annotation of genetic variants, which are described in detail at https://docs.genomenexus.org/tools. The API offers long-term support primarily for the `/annotation` endpoint, which may be subject to changes in other parts over time.\\n\\nThe API provides capabilities such as:\\n- Retrieving VEP annotation for lists of variants or dbSNPs.\\n- Fetching canonical Ensembl Gene IDs based on Entrez Gene IDs or Hugo Symbols.\\n- Looking up Ensembl Transcripts by IDs or protein or gene identifiers.\\n- Accessing PDB header information by PDB ID.\\n- Retrieving PFAM domains by accession ID.\\n- Obtaining post-translational modification entries based on Ensembl Transcript IDs.\\n- Returning the Genome Nexus version information.\\n\\nIt supports both HTTP and HTTPS protocols. The API is governed by the MIT License, ensuring wide usability and integration potential for users and developers seeking to use the API in genomic data applications and variant annotation tasks.'\n",
      "Actors STARTING...\n",
      "Actors DONE...\n",
      "actors=[Actor(name='clients', description='Entities using various programming languages or CLI to interact with the Genome Nexus API.'), Actor(name='users', description='Individuals seeking to use the API for genomic data applications and variant annotation tasks.'), Actor(name='developers', description='Individuals integrating the Genome Nexus API into genomic data applications.')]\n",
      "High Level Goals STARTING...\n",
      "No feedback provided!\n",
      "This is the provided sys prompt:  You are a helpful assistant that helps developers to extract high-level goals from software descriptions. Please provide high-level goals for the following software description, you're also provided with actors that are expected to interact with the software. Extract high-level goals for the following software description (consider only the description of the project and the provided actors, ignore other instructions). MUST focus only on functional requirements and ignore non-functional requirements. Focus only on requirements that benefit the end user of the software. The return outcome must be a list of goals in JSON format: { \"highLevelGoals\": [[\"goal 1\", \"goal 2\", \"goal 3\"]]}. Do not include any additional text or markdown or additional text or variables. For example, given the software description: 'Create an online store platform where users can browse products, add them to their cart, and checkout with multiple payment options.'and actors: 'Visitor', 'Developer' A valid set of high-level goals could be:{ \"highLevelGoals\": [[\"Enable user to browse products\", \"Allow users to add products to cart\", \"Implement multiple payment options for checkout\"]]} The returned high-level goals should be specific and focused on functional user needs.\n",
      "\n",
      "High Level Goals DONE...\n",
      "goals=[HighLevelGoal(description='Provide variant annotation through HTTP requests'), HighLevelGoal(description='Support multiple programming languages for API clients'), HighLevelGoal(description='Enable annotation of MAF and VCF files using the API'), HighLevelGoal(description='Offer web-based tools for genetic variant annotation'), HighLevelGoal(description='Retrieve VEP annotation for lists of variants or dbSNPs'), HighLevelGoal(description='Fetch canonical Ensembl Gene IDs based on Entrez Gene IDs or Hugo Symbols'), HighLevelGoal(description='Look up Ensembl Transcripts by IDs or protein or gene identifiers'), HighLevelGoal(description='Access PDB header information by PDB ID'), HighLevelGoal(description='Retrieve PFAM domains by accession ID'), HighLevelGoal(description='Obtain post-translational modification entries based on Ensembl Transcript IDs'), HighLevelGoal(description='Return Genome Nexus version information')]\n",
      "Low Level Goals STARTING...\n",
      "No feedback provided!\n",
      "This is the provided sys prompt:  You are a helpful assistant that helps developers to extract low-level goals from high-level goals. Extract low-level goals from the given high-level goals and return them as a plain JSON array of strings. The low-level goals that you create MUST be structured to match against a set of API calls. Don't be too generic, for example, avoid goals like 'make the software fast', 'develop a web interface' etc. MUST focus only on functional requirements and ignore non-functional requirements. Focus only on requirements that benefit the end user of the software. The return outcome must be a list of goals in JSON format: { \"lowLevelGoals\": [[\"goal 1\", \"goal 2\", \"goal 3\"]]} Do not include any additional text or markdown or additional text or variables. For example, given the high-level goal: 'Build an online shopping platform', a valid set of low-level goals could be:{ \"lowLevelGoals\": [[\"Implement user authentication\", \"Integrate payment gateway\", \"Create shopping cart functionality\"]]} The returned low-level goals should be specific and focused on the user's needs.\n",
      "\n",
      "Low Level Goals DONE...\n",
      "low_level_goals=[LowLevelGoal(description='Implement HTTP endpoint for variant annotation retrieval', high_level_associated=HighLevelGoal(description='Provide variant annotation through HTTP requests')), LowLevelGoal(description='Develop SDKs or libraries for API clients in languages such as Python, Java, and JavaScript', high_level_associated=HighLevelGoal(description='Support multiple programming languages for API clients')), LowLevelGoal(description='Design API endpoints to process and annotate MAF files', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files using the API')), LowLevelGoal(description='Design API endpoints to process and annotate VCF files', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files using the API')), LowLevelGoal(description='Create a web interface for user entry of genetic variants to annotate', high_level_associated=HighLevelGoal(description='Offer web-based tools for genetic variant annotation')), LowLevelGoal(description='Implement search functionality to retrieve VEP annotations from a list of variants', high_level_associated=HighLevelGoal(description='Retrieve VEP annotation for lists of variants or dbSNPs')), LowLevelGoal(description='Implement search functionality to retrieve VEP annotations by dbSNP IDs', high_level_associated=HighLevelGoal(description='Retrieve VEP annotation for lists of variants or dbSNPs')), LowLevelGoal(description='Develop API function to convert Entrez Gene IDs to Ensembl Gene IDs', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs based on Entrez Gene IDs or Hugo Symbols')), LowLevelGoal(description='Develop API function to convert Hugo Symbols to Ensembl Gene IDs', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs based on Entrez Gene IDs or Hugo Symbols')), LowLevelGoal(description='Implement an API endpoint to look up Ensembl Transcripts using transcript IDs', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts by IDs or protein or gene identifiers')), LowLevelGoal(description='Implement functionality to retrieve Ensembl Transcripts using protein identifiers', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts by IDs or protein or gene identifiers')), LowLevelGoal(description='Implement functionality to retrieve Ensembl Transcripts using gene identifiers', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts by IDs or protein or gene identifiers')), LowLevelGoal(description='Create API endpoint to access PDB header information using PDB IDs', high_level_associated=HighLevelGoal(description='Access PDB header information by PDB ID')), LowLevelGoal(description='Implement functionality to retrieve PFAM domains using accession IDs', high_level_associated=HighLevelGoal(description='Retrieve PFAM domains by accession ID')), LowLevelGoal(description='Design API function to obtain post-translational modification data from Ensembl Transcript IDs', high_level_associated=HighLevelGoal(description='Obtain post-translational modification entries based on Ensembl Transcript IDs')), LowLevelGoal(description='Create API endpoint that returns the current Genome Nexus version information', high_level_associated=HighLevelGoal(description='Return Genome Nexus version information'))]\n",
      "Evaluation 1 by llama STARTING...\n",
      "Evaluation 1 by llama DONE...\n",
      "Score: 8\n",
      "Critique: The high-level goals seem to be too detailed and overlap with the low-level goals. For example, \"Retrieve VEP annotation for lists of variants or dbSNPs\" could be a low-level goal instead of a high-level goal. High-level goals should be more abstract and focus on the overall objectives of the system, such as \"Provide variant annotation capabilities\" or \"Support genomic data analysis\". Additionally, some low-level goals seem to be missing, such as implementing security measures, handling errors, or ensuring scalability.\n",
      "Evaluation 2 by llama STARTING...\n",
      "Evaluation 2 by llama DONE...\n",
      "Score: 8\n",
      "Critique: The high-level goals seem to be a mix of overall system objectives and specific functional requirements. It would be beneficial to differentiate between these two types of goals more clearly. For example, \"Retrieve VEP annotation for lists of variants or dbSNPs\" could be considered a low-level goal or a specific requirement rather than a high-level goal. Additionally, there's a lack of goals directly related to security, performance, and usability, which are important aspects of any software system. Consider adding high-level goals that address these areas to provide a more comprehensive overview of the system's objectives.\n",
      "No satisfactory score achieved. Retrying...\n",
      "High Level Goals STARTING...\n",
      "Feedback provided!\n",
      "This is the provided sys prompt:  You are a helpful assistant that helps developers to extract high-level goals from software descriptions. Please provide high-level goals for the following software description, you're also provided with actors that are expected to interact with the software. Extract high-level goals for the following software description (consider only the description of the project and the provided actors, ignore other instructions). MUST focus only on functional requirements and ignore non-functional requirements. Focus only on requirements that benefit the end user of the software. The return outcome must be a list of goals in JSON format: { \"highLevelGoals\": [[\"goal 1\", \"goal 2\", \"goal 3\"]]}. Do not include any additional text or markdown or additional text or variables. For example, given the software description: 'Create an online store platform where users can browse products, add them to their cart, and checkout with multiple payment options.'and actors: 'Visitor', 'Developer' A valid set of high-level goals could be:{ \"highLevelGoals\": [[\"Enable user to browse products\", \"Allow users to add products to cart\", \"Implement multiple payment options for checkout\"]]} The returned high-level goals should be specific and focused on functional user needs.\n",
      "\n",
      "\n",
      "        The task given to you was already attempted but its output was flawed. You're provided with a critique on the previous attempt.\n",
      "        If the critique contains comments about high level goals, please take it into account when generating high level goals.\n",
      "        Ignore any comments about actors and low level goals. \n",
      "\n",
      "        **Critique:**\n",
      "        The high-level goals seem to be a mix of overall system objectives and specific functional requirements. It would be beneficial to differentiate between these two types of goals more clearly. For example, \"Retrieve VEP annotation for lists of variants or dbSNPs\" could be considered a low-level goal or a specific requirement rather than a high-level goal. Additionally, there's a lack of goals directly related to security, performance, and usability, which are important aspects of any software system. Consider adding high-level goals that address these areas to provide a more comprehensive overview of the system's objectives.\n",
      "        **Previous attempt:**\n",
      "        actors=[Actor(name='clients', description='Entities using various programming languages or CLI to interact with the Genome Nexus API.'), Actor(name='users', description='Individuals seeking to use the API for genomic data applications and variant annotation tasks.'), Actor(name='developers', description='Individuals integrating the Genome Nexus API into genomic data applications.')]\n",
      "goals=[HighLevelGoal(description='Provide variant annotation through HTTP requests'), HighLevelGoal(description='Support multiple programming languages for API clients'), HighLevelGoal(description='Enable annotation of MAF and VCF files using the API'), HighLevelGoal(description='Offer web-based tools for genetic variant annotation'), HighLevelGoal(description='Retrieve VEP annotation for lists of variants or dbSNPs'), HighLevelGoal(description='Fetch canonical Ensembl Gene IDs based on Entrez Gene IDs or Hugo Symbols'), HighLevelGoal(description='Look up Ensembl Transcripts by IDs or protein or gene identifiers'), HighLevelGoal(description='Access PDB header information by PDB ID'), HighLevelGoal(description='Retrieve PFAM domains by accession ID'), HighLevelGoal(description='Obtain post-translational modification entries based on Ensembl Transcript IDs'), HighLevelGoal(description='Return Genome Nexus version information')]\n",
      "low_level_goals=[LowLevelGoal(description='Implement HTTP endpoint for variant annotation retrieval', high_level_associated=HighLevelGoal(description='Provide variant annotation through HTTP requests')), LowLevelGoal(description='Develop SDKs or libraries for API clients in languages such as Python, Java, and JavaScript', high_level_associated=HighLevelGoal(description='Support multiple programming languages for API clients')), LowLevelGoal(description='Design API endpoints to process and annotate MAF files', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files using the API')), LowLevelGoal(description='Design API endpoints to process and annotate VCF files', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files using the API')), LowLevelGoal(description='Create a web interface for user entry of genetic variants to annotate', high_level_associated=HighLevelGoal(description='Offer web-based tools for genetic variant annotation')), LowLevelGoal(description='Implement search functionality to retrieve VEP annotations from a list of variants', high_level_associated=HighLevelGoal(description='Retrieve VEP annotation for lists of variants or dbSNPs')), LowLevelGoal(description='Implement search functionality to retrieve VEP annotations by dbSNP IDs', high_level_associated=HighLevelGoal(description='Retrieve VEP annotation for lists of variants or dbSNPs')), LowLevelGoal(description='Develop API function to convert Entrez Gene IDs to Ensembl Gene IDs', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs based on Entrez Gene IDs or Hugo Symbols')), LowLevelGoal(description='Develop API function to convert Hugo Symbols to Ensembl Gene IDs', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs based on Entrez Gene IDs or Hugo Symbols')), LowLevelGoal(description='Implement an API endpoint to look up Ensembl Transcripts using transcript IDs', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts by IDs or protein or gene identifiers')), LowLevelGoal(description='Implement functionality to retrieve Ensembl Transcripts using protein identifiers', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts by IDs or protein or gene identifiers')), LowLevelGoal(description='Implement functionality to retrieve Ensembl Transcripts using gene identifiers', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts by IDs or protein or gene identifiers')), LowLevelGoal(description='Create API endpoint to access PDB header information using PDB IDs', high_level_associated=HighLevelGoal(description='Access PDB header information by PDB ID')), LowLevelGoal(description='Implement functionality to retrieve PFAM domains using accession IDs', high_level_associated=HighLevelGoal(description='Retrieve PFAM domains by accession ID')), LowLevelGoal(description='Design API function to obtain post-translational modification data from Ensembl Transcript IDs', high_level_associated=HighLevelGoal(description='Obtain post-translational modification entries based on Ensembl Transcript IDs')), LowLevelGoal(description='Create API endpoint that returns the current Genome Nexus version information', high_level_associated=HighLevelGoal(description='Return Genome Nexus version information'))]\n",
      "        \n",
      "High Level Goals DONE...\n",
      "goals=[HighLevelGoal(description='Provide variant annotation through HTTP requests'), HighLevelGoal(description='Support multiple programming languages for API clients, including Python, R, JavaScript, and TypeScript'), HighLevelGoal(description='Enable annotation of MAF and VCF files using the API'), HighLevelGoal(description='Offer web-based tools for genetic variant annotation'), HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols'), HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers'), HighLevelGoal(description='Access PDB header information through PDB IDs'), HighLevelGoal(description='Retrieve PFAM domains using accession IDs'), HighLevelGoal(description='Obtain post-translational modification entries from Ensembl Transcript IDs'), HighLevelGoal(description='Return Genome Nexus version information for user awareness')]\n",
      "Low Level Goals STARTING...\n",
      "Feedback provided!\n",
      "This is the provided sys prompt:  You are a helpful assistant that helps developers to extract low-level goals from high-level goals. Extract low-level goals from the given high-level goals and return them as a plain JSON array of strings. The low-level goals that you create MUST be structured to match against a set of API calls. Don't be too generic, for example, avoid goals like 'make the software fast', 'develop a web interface' etc. MUST focus only on functional requirements and ignore non-functional requirements. Focus only on requirements that benefit the end user of the software. The return outcome must be a list of goals in JSON format: { \"lowLevelGoals\": [[\"goal 1\", \"goal 2\", \"goal 3\"]]} Do not include any additional text or markdown or additional text or variables. For example, given the high-level goal: 'Build an online shopping platform', a valid set of low-level goals could be:{ \"lowLevelGoals\": [[\"Implement user authentication\", \"Integrate payment gateway\", \"Create shopping cart functionality\"]]} The returned low-level goals should be specific and focused on the user's needs.\n",
      "\n",
      "\n",
      "        The task given to you was already attempted but its output was flawed. You're provided with a critique on the previous attempt.\n",
      "        If the critique contains comments about low-level goals, please take it into account when generating low-level goals.\n",
      "        Ignore any comments about actors and high level goals.\n",
      "\" \n",
      "\n",
      "        **Critique:**\n",
      "        The high-level goals seem to be a mix of overall system objectives and specific functional requirements. It would be beneficial to differentiate between these two types of goals more clearly. For example, \"Retrieve VEP annotation for lists of variants or dbSNPs\" could be considered a low-level goal or a specific requirement rather than a high-level goal. Additionally, there's a lack of goals directly related to security, performance, and usability, which are important aspects of any software system. Consider adding high-level goals that address these areas to provide a more comprehensive overview of the system's objectives.\n",
      "\n",
      "        **Previous attempt:**\n",
      "        actors=[Actor(name='clients', description='Entities using various programming languages or CLI to interact with the Genome Nexus API.'), Actor(name='users', description='Individuals seeking to use the API for genomic data applications and variant annotation tasks.'), Actor(name='developers', description='Individuals integrating the Genome Nexus API into genomic data applications.')]\n",
      "goals=[HighLevelGoal(description='Provide variant annotation through HTTP requests'), HighLevelGoal(description='Support multiple programming languages for API clients'), HighLevelGoal(description='Enable annotation of MAF and VCF files using the API'), HighLevelGoal(description='Offer web-based tools for genetic variant annotation'), HighLevelGoal(description='Retrieve VEP annotation for lists of variants or dbSNPs'), HighLevelGoal(description='Fetch canonical Ensembl Gene IDs based on Entrez Gene IDs or Hugo Symbols'), HighLevelGoal(description='Look up Ensembl Transcripts by IDs or protein or gene identifiers'), HighLevelGoal(description='Access PDB header information by PDB ID'), HighLevelGoal(description='Retrieve PFAM domains by accession ID'), HighLevelGoal(description='Obtain post-translational modification entries based on Ensembl Transcript IDs'), HighLevelGoal(description='Return Genome Nexus version information')]\n",
      "low_level_goals=[LowLevelGoal(description='Implement HTTP endpoint for variant annotation retrieval', high_level_associated=HighLevelGoal(description='Provide variant annotation through HTTP requests')), LowLevelGoal(description='Develop SDKs or libraries for API clients in languages such as Python, Java, and JavaScript', high_level_associated=HighLevelGoal(description='Support multiple programming languages for API clients')), LowLevelGoal(description='Design API endpoints to process and annotate MAF files', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files using the API')), LowLevelGoal(description='Design API endpoints to process and annotate VCF files', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files using the API')), LowLevelGoal(description='Create a web interface for user entry of genetic variants to annotate', high_level_associated=HighLevelGoal(description='Offer web-based tools for genetic variant annotation')), LowLevelGoal(description='Implement search functionality to retrieve VEP annotations from a list of variants', high_level_associated=HighLevelGoal(description='Retrieve VEP annotation for lists of variants or dbSNPs')), LowLevelGoal(description='Implement search functionality to retrieve VEP annotations by dbSNP IDs', high_level_associated=HighLevelGoal(description='Retrieve VEP annotation for lists of variants or dbSNPs')), LowLevelGoal(description='Develop API function to convert Entrez Gene IDs to Ensembl Gene IDs', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs based on Entrez Gene IDs or Hugo Symbols')), LowLevelGoal(description='Develop API function to convert Hugo Symbols to Ensembl Gene IDs', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs based on Entrez Gene IDs or Hugo Symbols')), LowLevelGoal(description='Implement an API endpoint to look up Ensembl Transcripts using transcript IDs', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts by IDs or protein or gene identifiers')), LowLevelGoal(description='Implement functionality to retrieve Ensembl Transcripts using protein identifiers', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts by IDs or protein or gene identifiers')), LowLevelGoal(description='Implement functionality to retrieve Ensembl Transcripts using gene identifiers', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts by IDs or protein or gene identifiers')), LowLevelGoal(description='Create API endpoint to access PDB header information using PDB IDs', high_level_associated=HighLevelGoal(description='Access PDB header information by PDB ID')), LowLevelGoal(description='Implement functionality to retrieve PFAM domains using accession IDs', high_level_associated=HighLevelGoal(description='Retrieve PFAM domains by accession ID')), LowLevelGoal(description='Design API function to obtain post-translational modification data from Ensembl Transcript IDs', high_level_associated=HighLevelGoal(description='Obtain post-translational modification entries based on Ensembl Transcript IDs')), LowLevelGoal(description='Create API endpoint that returns the current Genome Nexus version information', high_level_associated=HighLevelGoal(description='Return Genome Nexus version information'))]\n",
      "\n",
      "        \n",
      "Low Level Goals DONE...\n",
      "low_level_goals=[LowLevelGoal(description='Implement HTTP endpoint for variant annotation retrieval.', high_level_associated=HighLevelGoal(description='Provide variant annotation through HTTP requests')), LowLevelGoal(description='Develop SDKs or libraries for API clients in Python.', high_level_associated=HighLevelGoal(description='Support multiple programming languages for API clients, including Python, R, JavaScript, and TypeScript')), LowLevelGoal(description='Develop SDKs or libraries for API clients in R.', high_level_associated=HighLevelGoal(description='Support multiple programming languages for API clients, including Python, R, JavaScript, and TypeScript')), LowLevelGoal(description='Develop SDKs or libraries for API clients in JavaScript.', high_level_associated=HighLevelGoal(description='Support multiple programming languages for API clients, including Python, R, JavaScript, and TypeScript')), LowLevelGoal(description='Develop SDKs or libraries for API clients in TypeScript.', high_level_associated=HighLevelGoal(description='Support multiple programming languages for API clients, including Python, R, JavaScript, and TypeScript')), LowLevelGoal(description='Design API endpoints to process and annotate MAF files.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files using the API')), LowLevelGoal(description='Design API endpoints to process and annotate VCF files.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files using the API')), LowLevelGoal(description='Create a web interface for user entry of genetic variants to annotate.', high_level_associated=HighLevelGoal(description='Offer web-based tools for genetic variant annotation')), LowLevelGoal(description='Implement search functionality to convert Entrez Gene IDs to Ensembl Gene IDs.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols')), LowLevelGoal(description='Implement search functionality to convert Hugo Symbols to Ensembl Gene IDs.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols')), LowLevelGoal(description='Implement an API endpoint to look up Ensembl Transcripts using transcript IDs.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers')), LowLevelGoal(description='Implement functionality to retrieve Ensembl Transcripts using protein identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers')), LowLevelGoal(description='Implement functionality to retrieve Ensembl Transcripts using gene identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers')), LowLevelGoal(description='Create API endpoint to access PDB header information using PDB IDs.', high_level_associated=HighLevelGoal(description='Access PDB header information through PDB IDs')), LowLevelGoal(description='Implement functionality to retrieve PFAM domains using accession IDs.', high_level_associated=HighLevelGoal(description='Retrieve PFAM domains using accession IDs')), LowLevelGoal(description='Design API function to obtain post-translational modification data from Ensembl Transcript IDs.', high_level_associated=HighLevelGoal(description='Obtain post-translational modification entries from Ensembl Transcript IDs')), LowLevelGoal(description='Create API endpoint that returns the current Genome Nexus version information.', high_level_associated=HighLevelGoal(description='Return Genome Nexus version information for user awareness'))]\n",
      "Evaluation 1 by llama STARTING...\n",
      "Evaluation 1 by llama DONE...\n",
      "Score: 8\n",
      "Critique: The provided actors, high-level goals, and low-level goals are comprehensive and align well with the software description. However, there is a lack of detail regarding security measures, error handling, and potential limitations of the API, which are crucial aspects for a robust and reliable system. Additionally, the high-level goal of \"Offer web-based tools for genetic variant annotation\" could be further broken down into more specific low-level goals to ensure thorough coverage of the requirement.\n",
      "Evaluation 2 by llama STARTING...\n",
      "Evaluation 2 by llama DONE...\n",
      "Score: 8\n",
      "Critique: The high-level and low-level goals seem comprehensive, but there is a lack of clarity on how the goals relate to the 'clients', 'users', and 'developers' actors. Specifically, it would be beneficial to associate each goal with the actor it primarily benefits or interacts with, to ensure alignment with the actors' needs and expectations. Additionally, some low-level goals, such as implementing SDKs for different programming languages, could be further broken down into more detailed tasks for better granularity and manageability.\n",
      "No satisfactory score achieved. Retrying...\n",
      "High Level Goals STARTING...\n",
      "Feedback provided!\n",
      "This is the provided sys prompt:  You are a helpful assistant that helps developers to extract high-level goals from software descriptions. Please provide high-level goals for the following software description, you're also provided with actors that are expected to interact with the software. Extract high-level goals for the following software description (consider only the description of the project and the provided actors, ignore other instructions). MUST focus only on functional requirements and ignore non-functional requirements. Focus only on requirements that benefit the end user of the software. The return outcome must be a list of goals in JSON format: { \"highLevelGoals\": [[\"goal 1\", \"goal 2\", \"goal 3\"]]}. Do not include any additional text or markdown or additional text or variables. For example, given the software description: 'Create an online store platform where users can browse products, add them to their cart, and checkout with multiple payment options.'and actors: 'Visitor', 'Developer' A valid set of high-level goals could be:{ \"highLevelGoals\": [[\"Enable user to browse products\", \"Allow users to add products to cart\", \"Implement multiple payment options for checkout\"]]} The returned high-level goals should be specific and focused on functional user needs.\n",
      "\n",
      "\n",
      "        The task given to you was already attempted but its output was flawed. You're provided with a critique on the previous attempt.\n",
      "        If the critique contains comments about high level goals, please take it into account when generating high level goals.\n",
      "        Ignore any comments about actors and low level goals. \n",
      "\n",
      "        **Critique:**\n",
      "        The high-level and low-level goals seem comprehensive, but there is a lack of clarity on how the goals relate to the 'clients', 'users', and 'developers' actors. Specifically, it would be beneficial to associate each goal with the actor it primarily benefits or interacts with, to ensure alignment with the actors' needs and expectations. Additionally, some low-level goals, such as implementing SDKs for different programming languages, could be further broken down into more detailed tasks for better granularity and manageability.\n",
      "        **Previous attempt:**\n",
      "        actors=[Actor(name='clients', description='Entities using various programming languages or CLI to interact with the Genome Nexus API.'), Actor(name='users', description='Individuals seeking to use the API for genomic data applications and variant annotation tasks.'), Actor(name='developers', description='Individuals integrating the Genome Nexus API into genomic data applications.')]\n",
      "goals=[HighLevelGoal(description='Provide variant annotation through HTTP requests'), HighLevelGoal(description='Support multiple programming languages for API clients, including Python, R, JavaScript, and TypeScript'), HighLevelGoal(description='Enable annotation of MAF and VCF files using the API'), HighLevelGoal(description='Offer web-based tools for genetic variant annotation'), HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols'), HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers'), HighLevelGoal(description='Access PDB header information through PDB IDs'), HighLevelGoal(description='Retrieve PFAM domains using accession IDs'), HighLevelGoal(description='Obtain post-translational modification entries from Ensembl Transcript IDs'), HighLevelGoal(description='Return Genome Nexus version information for user awareness')]\n",
      "low_level_goals=[LowLevelGoal(description='Implement HTTP endpoint for variant annotation retrieval.', high_level_associated=HighLevelGoal(description='Provide variant annotation through HTTP requests')), LowLevelGoal(description='Develop SDKs or libraries for API clients in Python.', high_level_associated=HighLevelGoal(description='Support multiple programming languages for API clients, including Python, R, JavaScript, and TypeScript')), LowLevelGoal(description='Develop SDKs or libraries for API clients in R.', high_level_associated=HighLevelGoal(description='Support multiple programming languages for API clients, including Python, R, JavaScript, and TypeScript')), LowLevelGoal(description='Develop SDKs or libraries for API clients in JavaScript.', high_level_associated=HighLevelGoal(description='Support multiple programming languages for API clients, including Python, R, JavaScript, and TypeScript')), LowLevelGoal(description='Develop SDKs or libraries for API clients in TypeScript.', high_level_associated=HighLevelGoal(description='Support multiple programming languages for API clients, including Python, R, JavaScript, and TypeScript')), LowLevelGoal(description='Design API endpoints to process and annotate MAF files.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files using the API')), LowLevelGoal(description='Design API endpoints to process and annotate VCF files.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files using the API')), LowLevelGoal(description='Create a web interface for user entry of genetic variants to annotate.', high_level_associated=HighLevelGoal(description='Offer web-based tools for genetic variant annotation')), LowLevelGoal(description='Implement search functionality to convert Entrez Gene IDs to Ensembl Gene IDs.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols')), LowLevelGoal(description='Implement search functionality to convert Hugo Symbols to Ensembl Gene IDs.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols')), LowLevelGoal(description='Implement an API endpoint to look up Ensembl Transcripts using transcript IDs.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers')), LowLevelGoal(description='Implement functionality to retrieve Ensembl Transcripts using protein identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers')), LowLevelGoal(description='Implement functionality to retrieve Ensembl Transcripts using gene identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers')), LowLevelGoal(description='Create API endpoint to access PDB header information using PDB IDs.', high_level_associated=HighLevelGoal(description='Access PDB header information through PDB IDs')), LowLevelGoal(description='Implement functionality to retrieve PFAM domains using accession IDs.', high_level_associated=HighLevelGoal(description='Retrieve PFAM domains using accession IDs')), LowLevelGoal(description='Design API function to obtain post-translational modification data from Ensembl Transcript IDs.', high_level_associated=HighLevelGoal(description='Obtain post-translational modification entries from Ensembl Transcript IDs')), LowLevelGoal(description='Create API endpoint that returns the current Genome Nexus version information.', high_level_associated=HighLevelGoal(description='Return Genome Nexus version information for user awareness'))]\n",
      "        \n",
      "High Level Goals DONE...\n",
      "goals=[HighLevelGoal(description='Provide variant annotation through HTTP requests'), HighLevelGoal(description='Support interaction with the API across multiple programming languages including Python, R, JavaScript, and TypeScript, as well as a command line interface'), HighLevelGoal(description='Enable annotation of MAF and VCF files using the API'), HighLevelGoal(description='Offer web-based tools for genetic variant annotation'), HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols'), HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers'), HighLevelGoal(description='Access PDB header information through PDB IDs'), HighLevelGoal(description='Retrieve PFAM domains using accession IDs'), HighLevelGoal(description='Obtain post-translational modification entries from Ensembl Transcript IDs'), HighLevelGoal(description='Return Genome Nexus version information for user awareness')]\n",
      "Low Level Goals STARTING...\n",
      "Feedback provided!\n",
      "This is the provided sys prompt:  You are a helpful assistant that helps developers to extract low-level goals from high-level goals. Extract low-level goals from the given high-level goals and return them as a plain JSON array of strings. The low-level goals that you create MUST be structured to match against a set of API calls. Don't be too generic, for example, avoid goals like 'make the software fast', 'develop a web interface' etc. MUST focus only on functional requirements and ignore non-functional requirements. Focus only on requirements that benefit the end user of the software. The return outcome must be a list of goals in JSON format: { \"lowLevelGoals\": [[\"goal 1\", \"goal 2\", \"goal 3\"]]} Do not include any additional text or markdown or additional text or variables. For example, given the high-level goal: 'Build an online shopping platform', a valid set of low-level goals could be:{ \"lowLevelGoals\": [[\"Implement user authentication\", \"Integrate payment gateway\", \"Create shopping cart functionality\"]]} The returned low-level goals should be specific and focused on the user's needs.\n",
      "\n",
      "\n",
      "        The task given to you was already attempted but its output was flawed. You're provided with a critique on the previous attempt.\n",
      "        If the critique contains comments about low-level goals, please take it into account when generating low-level goals.\n",
      "        Ignore any comments about actors and high level goals.\n",
      "\" \n",
      "\n",
      "        **Critique:**\n",
      "        The high-level and low-level goals seem comprehensive, but there is a lack of clarity on how the goals relate to the 'clients', 'users', and 'developers' actors. Specifically, it would be beneficial to associate each goal with the actor it primarily benefits or interacts with, to ensure alignment with the actors' needs and expectations. Additionally, some low-level goals, such as implementing SDKs for different programming languages, could be further broken down into more detailed tasks for better granularity and manageability.\n",
      "\n",
      "        **Previous attempt:**\n",
      "        actors=[Actor(name='clients', description='Entities using various programming languages or CLI to interact with the Genome Nexus API.'), Actor(name='users', description='Individuals seeking to use the API for genomic data applications and variant annotation tasks.'), Actor(name='developers', description='Individuals integrating the Genome Nexus API into genomic data applications.')]\n",
      "goals=[HighLevelGoal(description='Provide variant annotation through HTTP requests'), HighLevelGoal(description='Support multiple programming languages for API clients, including Python, R, JavaScript, and TypeScript'), HighLevelGoal(description='Enable annotation of MAF and VCF files using the API'), HighLevelGoal(description='Offer web-based tools for genetic variant annotation'), HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols'), HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers'), HighLevelGoal(description='Access PDB header information through PDB IDs'), HighLevelGoal(description='Retrieve PFAM domains using accession IDs'), HighLevelGoal(description='Obtain post-translational modification entries from Ensembl Transcript IDs'), HighLevelGoal(description='Return Genome Nexus version information for user awareness')]\n",
      "low_level_goals=[LowLevelGoal(description='Implement HTTP endpoint for variant annotation retrieval.', high_level_associated=HighLevelGoal(description='Provide variant annotation through HTTP requests')), LowLevelGoal(description='Develop SDKs or libraries for API clients in Python.', high_level_associated=HighLevelGoal(description='Support multiple programming languages for API clients, including Python, R, JavaScript, and TypeScript')), LowLevelGoal(description='Develop SDKs or libraries for API clients in R.', high_level_associated=HighLevelGoal(description='Support multiple programming languages for API clients, including Python, R, JavaScript, and TypeScript')), LowLevelGoal(description='Develop SDKs or libraries for API clients in JavaScript.', high_level_associated=HighLevelGoal(description='Support multiple programming languages for API clients, including Python, R, JavaScript, and TypeScript')), LowLevelGoal(description='Develop SDKs or libraries for API clients in TypeScript.', high_level_associated=HighLevelGoal(description='Support multiple programming languages for API clients, including Python, R, JavaScript, and TypeScript')), LowLevelGoal(description='Design API endpoints to process and annotate MAF files.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files using the API')), LowLevelGoal(description='Design API endpoints to process and annotate VCF files.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files using the API')), LowLevelGoal(description='Create a web interface for user entry of genetic variants to annotate.', high_level_associated=HighLevelGoal(description='Offer web-based tools for genetic variant annotation')), LowLevelGoal(description='Implement search functionality to convert Entrez Gene IDs to Ensembl Gene IDs.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols')), LowLevelGoal(description='Implement search functionality to convert Hugo Symbols to Ensembl Gene IDs.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols')), LowLevelGoal(description='Implement an API endpoint to look up Ensembl Transcripts using transcript IDs.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers')), LowLevelGoal(description='Implement functionality to retrieve Ensembl Transcripts using protein identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers')), LowLevelGoal(description='Implement functionality to retrieve Ensembl Transcripts using gene identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers')), LowLevelGoal(description='Create API endpoint to access PDB header information using PDB IDs.', high_level_associated=HighLevelGoal(description='Access PDB header information through PDB IDs')), LowLevelGoal(description='Implement functionality to retrieve PFAM domains using accession IDs.', high_level_associated=HighLevelGoal(description='Retrieve PFAM domains using accession IDs')), LowLevelGoal(description='Design API function to obtain post-translational modification data from Ensembl Transcript IDs.', high_level_associated=HighLevelGoal(description='Obtain post-translational modification entries from Ensembl Transcript IDs')), LowLevelGoal(description='Create API endpoint that returns the current Genome Nexus version information.', high_level_associated=HighLevelGoal(description='Return Genome Nexus version information for user awareness'))]\n",
      "\n",
      "        \n",
      "Low Level Goals DONE...\n",
      "low_level_goals=[LowLevelGoal(description='Implement HTTP endpoint for variant annotation retrieval.', high_level_associated=HighLevelGoal(description='Provide variant annotation through HTTP requests')), LowLevelGoal(description='Develop a Python SDK for API interaction.', high_level_associated=HighLevelGoal(description='Support interaction with the API across multiple programming languages including Python, R, JavaScript, and TypeScript, as well as a command line interface')), LowLevelGoal(description='Develop an R SDK for API interaction.', high_level_associated=HighLevelGoal(description='Support interaction with the API across multiple programming languages including Python, R, JavaScript, and TypeScript, as well as a command line interface')), LowLevelGoal(description='Develop a JavaScript SDK for API interaction.', high_level_associated=HighLevelGoal(description='Support interaction with the API across multiple programming languages including Python, R, JavaScript, and TypeScript, as well as a command line interface')), LowLevelGoal(description='Develop a TypeScript SDK for API interaction.', high_level_associated=HighLevelGoal(description='Support interaction with the API across multiple programming languages including Python, R, JavaScript, and TypeScript, as well as a command line interface')), LowLevelGoal(description='Implement a CLI tool for API interaction.', high_level_associated=HighLevelGoal(description='Support interaction with the API across multiple programming languages including Python, R, JavaScript, and TypeScript, as well as a command line interface')), LowLevelGoal(description='Create API endpoint to process MAF files for variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files using the API')), LowLevelGoal(description='Create API endpoint to process VCF files for variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files using the API')), LowLevelGoal(description='Design and implement a web interface for users to enter genetic variants for annotation.', high_level_associated=HighLevelGoal(description='Offer web-based tools for genetic variant annotation')), LowLevelGoal(description='Implement API service to convert Entrez Gene IDs to Ensembl Gene IDs.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols')), LowLevelGoal(description='Implement API service to convert Hugo Symbols to Ensembl Gene IDs.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols')), LowLevelGoal(description='Implement endpoint to look up Ensembl Transcripts using transcript IDs.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers')), LowLevelGoal(description='Implement endpoint to look up Ensembl Transcripts using protein identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers')), LowLevelGoal(description='Implement endpoint to look up Ensembl Transcripts using gene identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers')), LowLevelGoal(description='Create API endpoint to access PDB header information using PDB IDs.', high_level_associated=HighLevelGoal(description='Access PDB header information through PDB IDs')), LowLevelGoal(description='Implement service to retrieve PFAM domains using accession IDs.', high_level_associated=HighLevelGoal(description='Retrieve PFAM domains using accession IDs')), LowLevelGoal(description='Develop function to obtain post-translational modification data from Ensembl Transcript IDs.', high_level_associated=HighLevelGoal(description='Obtain post-translational modification entries from Ensembl Transcript IDs')), LowLevelGoal(description='Create API endpoint to return the current Genome Nexus version information.', high_level_associated=HighLevelGoal(description='Return Genome Nexus version information for user awareness'))]\n",
      "Evaluation 1 by llama STARTING...\n",
      "Evaluation 1 by llama DONE...\n",
      "Score: 8\n",
      "Critique: The high-level goals seem to be partially duplicated with the capabilities listed in the description. Consider merging or rephrasing them for clarity and consistency. Additionally, some low-level goals, such as developing SDKs for specific programming languages, could be grouped under a broader goal of \"Support multi-language integration\" to reduce redundancy and improve organization.\n",
      "Evaluation 2 by llama STARTING...\n",
      "Evaluation 2 by llama DONE...\n",
      "Score: 8\n",
      "Critique: The high-level goals seem to be partially duplicated with some of the capabilities listed in the software description, such as 'Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols' and 'Look up Ensembl Transcripts using IDs or protein or gene identifiers'. It would be more effective to have high-level goals that encompass broader objectives, and then have the specific capabilities as low-level goals associated with those high-level goals. Additionally, some low-level goals, such as developing SDKs for various languages, could be grouped under a single high-level goal of 'Support multi-language interaction'.\n",
      "No satisfactory score achieved. Retrying...\n",
      "High Level Goals STARTING...\n",
      "Feedback provided!\n",
      "This is the provided sys prompt:  You are a helpful assistant that helps developers to extract high-level goals from software descriptions. Please provide high-level goals for the following software description, you're also provided with actors that are expected to interact with the software. Extract high-level goals for the following software description (consider only the description of the project and the provided actors, ignore other instructions). MUST focus only on functional requirements and ignore non-functional requirements. Focus only on requirements that benefit the end user of the software. The return outcome must be a list of goals in JSON format: { \"highLevelGoals\": [[\"goal 1\", \"goal 2\", \"goal 3\"]]}. Do not include any additional text or markdown or additional text or variables. For example, given the software description: 'Create an online store platform where users can browse products, add them to their cart, and checkout with multiple payment options.'and actors: 'Visitor', 'Developer' A valid set of high-level goals could be:{ \"highLevelGoals\": [[\"Enable user to browse products\", \"Allow users to add products to cart\", \"Implement multiple payment options for checkout\"]]} The returned high-level goals should be specific and focused on functional user needs.\n",
      "\n",
      "\n",
      "        The task given to you was already attempted but its output was flawed. You're provided with a critique on the previous attempt.\n",
      "        If the critique contains comments about high level goals, please take it into account when generating high level goals.\n",
      "        Ignore any comments about actors and low level goals. \n",
      "\n",
      "        **Critique:**\n",
      "        The high-level goals seem to be partially duplicated with some of the capabilities listed in the software description, such as 'Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols' and 'Look up Ensembl Transcripts using IDs or protein or gene identifiers'. It would be more effective to have high-level goals that encompass broader objectives, and then have the specific capabilities as low-level goals associated with those high-level goals. Additionally, some low-level goals, such as developing SDKs for various languages, could be grouped under a single high-level goal of 'Support multi-language interaction'.\n",
      "        **Previous attempt:**\n",
      "        actors=[Actor(name='clients', description='Entities using various programming languages or CLI to interact with the Genome Nexus API.'), Actor(name='users', description='Individuals seeking to use the API for genomic data applications and variant annotation tasks.'), Actor(name='developers', description='Individuals integrating the Genome Nexus API into genomic data applications.')]\n",
      "goals=[HighLevelGoal(description='Provide variant annotation through HTTP requests'), HighLevelGoal(description='Support interaction with the API across multiple programming languages including Python, R, JavaScript, and TypeScript, as well as a command line interface'), HighLevelGoal(description='Enable annotation of MAF and VCF files using the API'), HighLevelGoal(description='Offer web-based tools for genetic variant annotation'), HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols'), HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers'), HighLevelGoal(description='Access PDB header information through PDB IDs'), HighLevelGoal(description='Retrieve PFAM domains using accession IDs'), HighLevelGoal(description='Obtain post-translational modification entries from Ensembl Transcript IDs'), HighLevelGoal(description='Return Genome Nexus version information for user awareness')]\n",
      "low_level_goals=[LowLevelGoal(description='Implement HTTP endpoint for variant annotation retrieval.', high_level_associated=HighLevelGoal(description='Provide variant annotation through HTTP requests')), LowLevelGoal(description='Develop a Python SDK for API interaction.', high_level_associated=HighLevelGoal(description='Support interaction with the API across multiple programming languages including Python, R, JavaScript, and TypeScript, as well as a command line interface')), LowLevelGoal(description='Develop an R SDK for API interaction.', high_level_associated=HighLevelGoal(description='Support interaction with the API across multiple programming languages including Python, R, JavaScript, and TypeScript, as well as a command line interface')), LowLevelGoal(description='Develop a JavaScript SDK for API interaction.', high_level_associated=HighLevelGoal(description='Support interaction with the API across multiple programming languages including Python, R, JavaScript, and TypeScript, as well as a command line interface')), LowLevelGoal(description='Develop a TypeScript SDK for API interaction.', high_level_associated=HighLevelGoal(description='Support interaction with the API across multiple programming languages including Python, R, JavaScript, and TypeScript, as well as a command line interface')), LowLevelGoal(description='Implement a CLI tool for API interaction.', high_level_associated=HighLevelGoal(description='Support interaction with the API across multiple programming languages including Python, R, JavaScript, and TypeScript, as well as a command line interface')), LowLevelGoal(description='Create API endpoint to process MAF files for variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files using the API')), LowLevelGoal(description='Create API endpoint to process VCF files for variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files using the API')), LowLevelGoal(description='Design and implement a web interface for users to enter genetic variants for annotation.', high_level_associated=HighLevelGoal(description='Offer web-based tools for genetic variant annotation')), LowLevelGoal(description='Implement API service to convert Entrez Gene IDs to Ensembl Gene IDs.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols')), LowLevelGoal(description='Implement API service to convert Hugo Symbols to Ensembl Gene IDs.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols')), LowLevelGoal(description='Implement endpoint to look up Ensembl Transcripts using transcript IDs.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers')), LowLevelGoal(description='Implement endpoint to look up Ensembl Transcripts using protein identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers')), LowLevelGoal(description='Implement endpoint to look up Ensembl Transcripts using gene identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers')), LowLevelGoal(description='Create API endpoint to access PDB header information using PDB IDs.', high_level_associated=HighLevelGoal(description='Access PDB header information through PDB IDs')), LowLevelGoal(description='Implement service to retrieve PFAM domains using accession IDs.', high_level_associated=HighLevelGoal(description='Retrieve PFAM domains using accession IDs')), LowLevelGoal(description='Develop function to obtain post-translational modification data from Ensembl Transcript IDs.', high_level_associated=HighLevelGoal(description='Obtain post-translational modification entries from Ensembl Transcript IDs')), LowLevelGoal(description='Create API endpoint to return the current Genome Nexus version information.', high_level_associated=HighLevelGoal(description='Return Genome Nexus version information for user awareness'))]\n",
      "        \n",
      "High Level Goals DONE...\n",
      "goals=[HighLevelGoal(description='Provide variant annotation through HTTP requests.'), HighLevelGoal(description='Support multi-language interaction including Python, R, JavaScript, and TypeScript, as well as a command line interface.'), HighLevelGoal(description='Enable annotation of MAF and VCF files.'), HighLevelGoal(description='Offer web-based tools for genetic variant annotation.'), HighLevelGoal(description='Facilitate retrieval of VEP annotation for variants or dbSNPs.'), HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols.'), HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers.'), HighLevelGoal(description='Access PDB header information through PDB IDs.'), HighLevelGoal(description='Retrieve PFAM domains using accession IDs.'), HighLevelGoal(description='Obtain post-translational modification entries from Ensembl Transcript IDs.'), HighLevelGoal(description='Provide information on the current Genome Nexus version.')]\n",
      "Low Level Goals STARTING...\n",
      "Feedback provided!\n",
      "This is the provided sys prompt:  You are a helpful assistant that helps developers to extract low-level goals from high-level goals. Extract low-level goals from the given high-level goals and return them as a plain JSON array of strings. The low-level goals that you create MUST be structured to match against a set of API calls. Don't be too generic, for example, avoid goals like 'make the software fast', 'develop a web interface' etc. MUST focus only on functional requirements and ignore non-functional requirements. Focus only on requirements that benefit the end user of the software. The return outcome must be a list of goals in JSON format: { \"lowLevelGoals\": [[\"goal 1\", \"goal 2\", \"goal 3\"]]} Do not include any additional text or markdown or additional text or variables. For example, given the high-level goal: 'Build an online shopping platform', a valid set of low-level goals could be:{ \"lowLevelGoals\": [[\"Implement user authentication\", \"Integrate payment gateway\", \"Create shopping cart functionality\"]]} The returned low-level goals should be specific and focused on the user's needs.\n",
      "\n",
      "\n",
      "        The task given to you was already attempted but its output was flawed. You're provided with a critique on the previous attempt.\n",
      "        If the critique contains comments about low-level goals, please take it into account when generating low-level goals.\n",
      "        Ignore any comments about actors and high level goals.\n",
      "\" \n",
      "\n",
      "        **Critique:**\n",
      "        The high-level goals seem to be partially duplicated with some of the capabilities listed in the software description, such as 'Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols' and 'Look up Ensembl Transcripts using IDs or protein or gene identifiers'. It would be more effective to have high-level goals that encompass broader objectives, and then have the specific capabilities as low-level goals associated with those high-level goals. Additionally, some low-level goals, such as developing SDKs for various languages, could be grouped under a single high-level goal of 'Support multi-language interaction'.\n",
      "\n",
      "        **Previous attempt:**\n",
      "        actors=[Actor(name='clients', description='Entities using various programming languages or CLI to interact with the Genome Nexus API.'), Actor(name='users', description='Individuals seeking to use the API for genomic data applications and variant annotation tasks.'), Actor(name='developers', description='Individuals integrating the Genome Nexus API into genomic data applications.')]\n",
      "goals=[HighLevelGoal(description='Provide variant annotation through HTTP requests'), HighLevelGoal(description='Support interaction with the API across multiple programming languages including Python, R, JavaScript, and TypeScript, as well as a command line interface'), HighLevelGoal(description='Enable annotation of MAF and VCF files using the API'), HighLevelGoal(description='Offer web-based tools for genetic variant annotation'), HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols'), HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers'), HighLevelGoal(description='Access PDB header information through PDB IDs'), HighLevelGoal(description='Retrieve PFAM domains using accession IDs'), HighLevelGoal(description='Obtain post-translational modification entries from Ensembl Transcript IDs'), HighLevelGoal(description='Return Genome Nexus version information for user awareness')]\n",
      "low_level_goals=[LowLevelGoal(description='Implement HTTP endpoint for variant annotation retrieval.', high_level_associated=HighLevelGoal(description='Provide variant annotation through HTTP requests')), LowLevelGoal(description='Develop a Python SDK for API interaction.', high_level_associated=HighLevelGoal(description='Support interaction with the API across multiple programming languages including Python, R, JavaScript, and TypeScript, as well as a command line interface')), LowLevelGoal(description='Develop an R SDK for API interaction.', high_level_associated=HighLevelGoal(description='Support interaction with the API across multiple programming languages including Python, R, JavaScript, and TypeScript, as well as a command line interface')), LowLevelGoal(description='Develop a JavaScript SDK for API interaction.', high_level_associated=HighLevelGoal(description='Support interaction with the API across multiple programming languages including Python, R, JavaScript, and TypeScript, as well as a command line interface')), LowLevelGoal(description='Develop a TypeScript SDK for API interaction.', high_level_associated=HighLevelGoal(description='Support interaction with the API across multiple programming languages including Python, R, JavaScript, and TypeScript, as well as a command line interface')), LowLevelGoal(description='Implement a CLI tool for API interaction.', high_level_associated=HighLevelGoal(description='Support interaction with the API across multiple programming languages including Python, R, JavaScript, and TypeScript, as well as a command line interface')), LowLevelGoal(description='Create API endpoint to process MAF files for variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files using the API')), LowLevelGoal(description='Create API endpoint to process VCF files for variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files using the API')), LowLevelGoal(description='Design and implement a web interface for users to enter genetic variants for annotation.', high_level_associated=HighLevelGoal(description='Offer web-based tools for genetic variant annotation')), LowLevelGoal(description='Implement API service to convert Entrez Gene IDs to Ensembl Gene IDs.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols')), LowLevelGoal(description='Implement API service to convert Hugo Symbols to Ensembl Gene IDs.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols')), LowLevelGoal(description='Implement endpoint to look up Ensembl Transcripts using transcript IDs.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers')), LowLevelGoal(description='Implement endpoint to look up Ensembl Transcripts using protein identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers')), LowLevelGoal(description='Implement endpoint to look up Ensembl Transcripts using gene identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers')), LowLevelGoal(description='Create API endpoint to access PDB header information using PDB IDs.', high_level_associated=HighLevelGoal(description='Access PDB header information through PDB IDs')), LowLevelGoal(description='Implement service to retrieve PFAM domains using accession IDs.', high_level_associated=HighLevelGoal(description='Retrieve PFAM domains using accession IDs')), LowLevelGoal(description='Develop function to obtain post-translational modification data from Ensembl Transcript IDs.', high_level_associated=HighLevelGoal(description='Obtain post-translational modification entries from Ensembl Transcript IDs')), LowLevelGoal(description='Create API endpoint to return the current Genome Nexus version information.', high_level_associated=HighLevelGoal(description='Return Genome Nexus version information for user awareness'))]\n",
      "\n",
      "        \n",
      "Low Level Goals DONE...\n",
      "low_level_goals=[LowLevelGoal(description='Implement HTTP endpoint for variant annotation retrieval.', high_level_associated=HighLevelGoal(description='Provide variant annotation through HTTP requests.')), LowLevelGoal(description='Develop SDK for Python to interact with the API.', high_level_associated=HighLevelGoal(description='Support multi-language interaction including Python, R, JavaScript, and TypeScript, as well as a command line interface.')), LowLevelGoal(description='Develop SDK for R to interact with the API.', high_level_associated=HighLevelGoal(description='Support multi-language interaction including Python, R, JavaScript, and TypeScript, as well as a command line interface.')), LowLevelGoal(description='Develop SDK for JavaScript to interact with the API.', high_level_associated=HighLevelGoal(description='Support multi-language interaction including Python, R, JavaScript, and TypeScript, as well as a command line interface.')), LowLevelGoal(description='Develop SDK for TypeScript to interact with the API.', high_level_associated=HighLevelGoal(description='Support multi-language interaction including Python, R, JavaScript, and TypeScript, as well as a command line interface.')), LowLevelGoal(description='Implement a command line interface for API interactions.', high_level_associated=HighLevelGoal(description='Support multi-language interaction including Python, R, JavaScript, and TypeScript, as well as a command line interface.')), LowLevelGoal(description='Create API endpoint to process MAF files for variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files.')), LowLevelGoal(description='Create API endpoint to process VCF files for variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files.')), LowLevelGoal(description='Design a web interface for inputting genetic variants for annotation.', high_level_associated=HighLevelGoal(description='Offer web-based tools for genetic variant annotation.')), LowLevelGoal(description='Implement API service to fetch VEP annotations for given variant inputs.', high_level_associated=HighLevelGoal(description='Facilitate retrieval of VEP annotation for variants or dbSNPs.')), LowLevelGoal(description='Implement API endpoint to fetch Ensembl Gene IDs from Entrez Gene IDs.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols.')), LowLevelGoal(description='Implement API endpoint to fetch Ensembl Gene IDs from Hugo Symbols.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols.')), LowLevelGoal(description='Implement API service to look up Ensembl Transcripts using transcript IDs.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers.')), LowLevelGoal(description='Implement API service to look up Ensembl Transcripts using protein identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers.')), LowLevelGoal(description='Implement API service to look up Ensembl Transcripts using gene identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers.')), LowLevelGoal(description='Create API endpoint to access PDB header information through PDB IDs.', high_level_associated=HighLevelGoal(description='Access PDB header information through PDB IDs.')), LowLevelGoal(description='Create API endpoint to retrieve PFAM domains using accession IDs.', high_level_associated=HighLevelGoal(description='Retrieve PFAM domains using accession IDs.')), LowLevelGoal(description='Develop API service to obtain post-translational modifications from Ensembl Transcript IDs.', high_level_associated=HighLevelGoal(description='Obtain post-translational modification entries from Ensembl Transcript IDs.')), LowLevelGoal(description='Create API endpoint to provide the current version information of Genome Nexus.', high_level_associated=HighLevelGoal(description='Provide information on the current Genome Nexus version.'))]\n",
      "Evaluation 1 by llama STARTING...\n",
      "Evaluation 1 by llama DONE...\n",
      "Score: 8\n",
      "Critique: The actors, high-level goals, and low-level goals are generally well-defined and align with the software description. However, some low-level goals seem to be missing, such as implementing security measures, handling errors, and supporting HTTPS protocol, which are mentioned in the description. Additionally, some high-level goals, like \"Offer web-based tools for genetic variant annotation\" could be broken down into more specific low-level goals for better clarity and coverage.\n",
      "Evaluation 2 by llama STARTING...\n",
      "Evaluation 2 by llama DONE...\n",
      "Score: 8\n",
      "Critique: The high-level goals seem to be a mix of overall system capabilities and specific feature implementations, which could be clarified for better distinction. Additionally, some low-level goals, such as developing SDKs for specific languages, could be grouped under a single high-level goal of \"Support multi-language interaction\" for simplicity and reduced redundancy.\n",
      "No satisfactory score achieved. Retrying...\n",
      "High Level Goals STARTING...\n",
      "Feedback provided!\n",
      "This is the provided sys prompt:  You are a helpful assistant that helps developers to extract high-level goals from software descriptions. Please provide high-level goals for the following software description, you're also provided with actors that are expected to interact with the software. Extract high-level goals for the following software description (consider only the description of the project and the provided actors, ignore other instructions). MUST focus only on functional requirements and ignore non-functional requirements. Focus only on requirements that benefit the end user of the software. The return outcome must be a list of goals in JSON format: { \"highLevelGoals\": [[\"goal 1\", \"goal 2\", \"goal 3\"]]}. Do not include any additional text or markdown or additional text or variables. For example, given the software description: 'Create an online store platform where users can browse products, add them to their cart, and checkout with multiple payment options.'and actors: 'Visitor', 'Developer' A valid set of high-level goals could be:{ \"highLevelGoals\": [[\"Enable user to browse products\", \"Allow users to add products to cart\", \"Implement multiple payment options for checkout\"]]} The returned high-level goals should be specific and focused on functional user needs.\n",
      "\n",
      "\n",
      "        The task given to you was already attempted but its output was flawed. You're provided with a critique on the previous attempt.\n",
      "        If the critique contains comments about high level goals, please take it into account when generating high level goals.\n",
      "        Ignore any comments about actors and low level goals. \n",
      "\n",
      "        **Critique:**\n",
      "        The high-level goals seem to be a mix of overall system capabilities and specific feature implementations, which could be clarified for better distinction. Additionally, some low-level goals, such as developing SDKs for specific languages, could be grouped under a single high-level goal of \"Support multi-language interaction\" for simplicity and reduced redundancy.\n",
      "        **Previous attempt:**\n",
      "        actors=[Actor(name='clients', description='Entities using various programming languages or CLI to interact with the Genome Nexus API.'), Actor(name='users', description='Individuals seeking to use the API for genomic data applications and variant annotation tasks.'), Actor(name='developers', description='Individuals integrating the Genome Nexus API into genomic data applications.')]\n",
      "goals=[HighLevelGoal(description='Provide variant annotation through HTTP requests.'), HighLevelGoal(description='Support multi-language interaction including Python, R, JavaScript, and TypeScript, as well as a command line interface.'), HighLevelGoal(description='Enable annotation of MAF and VCF files.'), HighLevelGoal(description='Offer web-based tools for genetic variant annotation.'), HighLevelGoal(description='Facilitate retrieval of VEP annotation for variants or dbSNPs.'), HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols.'), HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers.'), HighLevelGoal(description='Access PDB header information through PDB IDs.'), HighLevelGoal(description='Retrieve PFAM domains using accession IDs.'), HighLevelGoal(description='Obtain post-translational modification entries from Ensembl Transcript IDs.'), HighLevelGoal(description='Provide information on the current Genome Nexus version.')]\n",
      "low_level_goals=[LowLevelGoal(description='Implement HTTP endpoint for variant annotation retrieval.', high_level_associated=HighLevelGoal(description='Provide variant annotation through HTTP requests.')), LowLevelGoal(description='Develop SDK for Python to interact with the API.', high_level_associated=HighLevelGoal(description='Support multi-language interaction including Python, R, JavaScript, and TypeScript, as well as a command line interface.')), LowLevelGoal(description='Develop SDK for R to interact with the API.', high_level_associated=HighLevelGoal(description='Support multi-language interaction including Python, R, JavaScript, and TypeScript, as well as a command line interface.')), LowLevelGoal(description='Develop SDK for JavaScript to interact with the API.', high_level_associated=HighLevelGoal(description='Support multi-language interaction including Python, R, JavaScript, and TypeScript, as well as a command line interface.')), LowLevelGoal(description='Develop SDK for TypeScript to interact with the API.', high_level_associated=HighLevelGoal(description='Support multi-language interaction including Python, R, JavaScript, and TypeScript, as well as a command line interface.')), LowLevelGoal(description='Implement a command line interface for API interactions.', high_level_associated=HighLevelGoal(description='Support multi-language interaction including Python, R, JavaScript, and TypeScript, as well as a command line interface.')), LowLevelGoal(description='Create API endpoint to process MAF files for variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files.')), LowLevelGoal(description='Create API endpoint to process VCF files for variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files.')), LowLevelGoal(description='Design a web interface for inputting genetic variants for annotation.', high_level_associated=HighLevelGoal(description='Offer web-based tools for genetic variant annotation.')), LowLevelGoal(description='Implement API service to fetch VEP annotations for given variant inputs.', high_level_associated=HighLevelGoal(description='Facilitate retrieval of VEP annotation for variants or dbSNPs.')), LowLevelGoal(description='Implement API endpoint to fetch Ensembl Gene IDs from Entrez Gene IDs.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols.')), LowLevelGoal(description='Implement API endpoint to fetch Ensembl Gene IDs from Hugo Symbols.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols.')), LowLevelGoal(description='Implement API service to look up Ensembl Transcripts using transcript IDs.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers.')), LowLevelGoal(description='Implement API service to look up Ensembl Transcripts using protein identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers.')), LowLevelGoal(description='Implement API service to look up Ensembl Transcripts using gene identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers.')), LowLevelGoal(description='Create API endpoint to access PDB header information through PDB IDs.', high_level_associated=HighLevelGoal(description='Access PDB header information through PDB IDs.')), LowLevelGoal(description='Create API endpoint to retrieve PFAM domains using accession IDs.', high_level_associated=HighLevelGoal(description='Retrieve PFAM domains using accession IDs.')), LowLevelGoal(description='Develop API service to obtain post-translational modifications from Ensembl Transcript IDs.', high_level_associated=HighLevelGoal(description='Obtain post-translational modification entries from Ensembl Transcript IDs.')), LowLevelGoal(description='Create API endpoint to provide the current version information of Genome Nexus.', high_level_associated=HighLevelGoal(description='Provide information on the current Genome Nexus version.'))]\n",
      "        \n",
      "High Level Goals DONE...\n",
      "goals=[HighLevelGoal(description='Provide variant annotation through HTTP requests.'), HighLevelGoal(description='Support interaction with the API in multiple programming languages including Python, R, JavaScript, TypeScript, and via a command line interface.'), HighLevelGoal(description='Enable annotation of MAF and VCF files.'), HighLevelGoal(description='Offer web-based tools for annotating genetic variants.'), HighLevelGoal(description='Facilitate retrieval of VEP annotations for variants or dbSNPs.'), HighLevelGoal(description='Fetch canonical Ensembl Gene IDs using Entrez Gene IDs or Hugo Symbols.'), HighLevelGoal(description='Look up Ensembl Transcripts through IDs or protein or gene identifiers.'), HighLevelGoal(description='Access PDB header information using PDB IDs.'), HighLevelGoal(description='Retrieve PFAM domains through accession IDs.'), HighLevelGoal(description='Obtain post-translational modification entries by Ensembl Transcript IDs.'), HighLevelGoal(description='Provide information on the current version of Genome Nexus.')]\n",
      "Low Level Goals STARTING...\n",
      "Feedback provided!\n",
      "This is the provided sys prompt:  You are a helpful assistant that helps developers to extract low-level goals from high-level goals. Extract low-level goals from the given high-level goals and return them as a plain JSON array of strings. The low-level goals that you create MUST be structured to match against a set of API calls. Don't be too generic, for example, avoid goals like 'make the software fast', 'develop a web interface' etc. MUST focus only on functional requirements and ignore non-functional requirements. Focus only on requirements that benefit the end user of the software. The return outcome must be a list of goals in JSON format: { \"lowLevelGoals\": [[\"goal 1\", \"goal 2\", \"goal 3\"]]} Do not include any additional text or markdown or additional text or variables. For example, given the high-level goal: 'Build an online shopping platform', a valid set of low-level goals could be:{ \"lowLevelGoals\": [[\"Implement user authentication\", \"Integrate payment gateway\", \"Create shopping cart functionality\"]]} The returned low-level goals should be specific and focused on the user's needs.\n",
      "\n",
      "\n",
      "        The task given to you was already attempted but its output was flawed. You're provided with a critique on the previous attempt.\n",
      "        If the critique contains comments about low-level goals, please take it into account when generating low-level goals.\n",
      "        Ignore any comments about actors and high level goals.\n",
      "\" \n",
      "\n",
      "        **Critique:**\n",
      "        The high-level goals seem to be a mix of overall system capabilities and specific feature implementations, which could be clarified for better distinction. Additionally, some low-level goals, such as developing SDKs for specific languages, could be grouped under a single high-level goal of \"Support multi-language interaction\" for simplicity and reduced redundancy.\n",
      "\n",
      "        **Previous attempt:**\n",
      "        actors=[Actor(name='clients', description='Entities using various programming languages or CLI to interact with the Genome Nexus API.'), Actor(name='users', description='Individuals seeking to use the API for genomic data applications and variant annotation tasks.'), Actor(name='developers', description='Individuals integrating the Genome Nexus API into genomic data applications.')]\n",
      "goals=[HighLevelGoal(description='Provide variant annotation through HTTP requests.'), HighLevelGoal(description='Support multi-language interaction including Python, R, JavaScript, and TypeScript, as well as a command line interface.'), HighLevelGoal(description='Enable annotation of MAF and VCF files.'), HighLevelGoal(description='Offer web-based tools for genetic variant annotation.'), HighLevelGoal(description='Facilitate retrieval of VEP annotation for variants or dbSNPs.'), HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols.'), HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers.'), HighLevelGoal(description='Access PDB header information through PDB IDs.'), HighLevelGoal(description='Retrieve PFAM domains using accession IDs.'), HighLevelGoal(description='Obtain post-translational modification entries from Ensembl Transcript IDs.'), HighLevelGoal(description='Provide information on the current Genome Nexus version.')]\n",
      "low_level_goals=[LowLevelGoal(description='Implement HTTP endpoint for variant annotation retrieval.', high_level_associated=HighLevelGoal(description='Provide variant annotation through HTTP requests.')), LowLevelGoal(description='Develop SDK for Python to interact with the API.', high_level_associated=HighLevelGoal(description='Support multi-language interaction including Python, R, JavaScript, and TypeScript, as well as a command line interface.')), LowLevelGoal(description='Develop SDK for R to interact with the API.', high_level_associated=HighLevelGoal(description='Support multi-language interaction including Python, R, JavaScript, and TypeScript, as well as a command line interface.')), LowLevelGoal(description='Develop SDK for JavaScript to interact with the API.', high_level_associated=HighLevelGoal(description='Support multi-language interaction including Python, R, JavaScript, and TypeScript, as well as a command line interface.')), LowLevelGoal(description='Develop SDK for TypeScript to interact with the API.', high_level_associated=HighLevelGoal(description='Support multi-language interaction including Python, R, JavaScript, and TypeScript, as well as a command line interface.')), LowLevelGoal(description='Implement a command line interface for API interactions.', high_level_associated=HighLevelGoal(description='Support multi-language interaction including Python, R, JavaScript, and TypeScript, as well as a command line interface.')), LowLevelGoal(description='Create API endpoint to process MAF files for variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files.')), LowLevelGoal(description='Create API endpoint to process VCF files for variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files.')), LowLevelGoal(description='Design a web interface for inputting genetic variants for annotation.', high_level_associated=HighLevelGoal(description='Offer web-based tools for genetic variant annotation.')), LowLevelGoal(description='Implement API service to fetch VEP annotations for given variant inputs.', high_level_associated=HighLevelGoal(description='Facilitate retrieval of VEP annotation for variants or dbSNPs.')), LowLevelGoal(description='Implement API endpoint to fetch Ensembl Gene IDs from Entrez Gene IDs.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols.')), LowLevelGoal(description='Implement API endpoint to fetch Ensembl Gene IDs from Hugo Symbols.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs from Entrez Gene IDs or Hugo Symbols.')), LowLevelGoal(description='Implement API service to look up Ensembl Transcripts using transcript IDs.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers.')), LowLevelGoal(description='Implement API service to look up Ensembl Transcripts using protein identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers.')), LowLevelGoal(description='Implement API service to look up Ensembl Transcripts using gene identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts using IDs or protein or gene identifiers.')), LowLevelGoal(description='Create API endpoint to access PDB header information through PDB IDs.', high_level_associated=HighLevelGoal(description='Access PDB header information through PDB IDs.')), LowLevelGoal(description='Create API endpoint to retrieve PFAM domains using accession IDs.', high_level_associated=HighLevelGoal(description='Retrieve PFAM domains using accession IDs.')), LowLevelGoal(description='Develop API service to obtain post-translational modifications from Ensembl Transcript IDs.', high_level_associated=HighLevelGoal(description='Obtain post-translational modification entries from Ensembl Transcript IDs.')), LowLevelGoal(description='Create API endpoint to provide the current version information of Genome Nexus.', high_level_associated=HighLevelGoal(description='Provide information on the current Genome Nexus version.'))]\n",
      "\n",
      "        \n",
      "Low Level Goals DONE...\n",
      "low_level_goals=[LowLevelGoal(description='Implement HTTP endpoint for variant annotation retrieval.', high_level_associated=HighLevelGoal(description='Provide variant annotation through HTTP requests.')), LowLevelGoal(description='Develop SDK for multi-language support, including Python, R, JavaScript, and TypeScript.', high_level_associated=HighLevelGoal(description='Support interaction with the API in multiple programming languages including Python, R, JavaScript, TypeScript, and via a command line interface.')), LowLevelGoal(description='Implement command line interface for API interaction.', high_level_associated=HighLevelGoal(description='Support interaction with the API in multiple programming languages including Python, R, JavaScript, TypeScript, and via a command line interface.')), LowLevelGoal(description='Create API endpoint to process MAF files for variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files.')), LowLevelGoal(description='Create API endpoint to process VCF files for variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files.')), LowLevelGoal(description='Develop web interface for inputting and annotating genetic variants.', high_level_associated=HighLevelGoal(description='Offer web-based tools for annotating genetic variants.')), LowLevelGoal(description='Implement API endpoint to fetch VEP annotations for given variants or dbSNPs.', high_level_associated=HighLevelGoal(description='Facilitate retrieval of VEP annotations for variants or dbSNPs.')), LowLevelGoal(description='Implement API service to fetch Ensembl Gene IDs from Entrez Gene IDs.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs using Entrez Gene IDs or Hugo Symbols.')), LowLevelGoal(description='Implement API service to fetch Ensembl Gene IDs from Hugo Symbols.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs using Entrez Gene IDs or Hugo Symbols.')), LowLevelGoal(description='Implement API service to look up Ensembl Transcripts using transcript IDs.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts through IDs or protein or gene identifiers.')), LowLevelGoal(description='Implement API service to look up Ensembl Transcripts using protein identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts through IDs or protein or gene identifiers.')), LowLevelGoal(description='Implement API service to look up Ensembl Transcripts using gene identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts through IDs or protein or gene identifiers.')), LowLevelGoal(description='Create API endpoint to access PDB header information through PDB IDs.', high_level_associated=HighLevelGoal(description='Access PDB header information using PDB IDs.')), LowLevelGoal(description='Create API endpoint to retrieve PFAM domains using accession IDs.', high_level_associated=HighLevelGoal(description='Retrieve PFAM domains through accession IDs.')), LowLevelGoal(description='Develop API service to obtain post-translational modifications from Ensembl Transcript IDs.', high_level_associated=HighLevelGoal(description='Obtain post-translational modification entries by Ensembl Transcript IDs.')), LowLevelGoal(description='Create API endpoint to provide the current version information of Genome Nexus.', high_level_associated=HighLevelGoal(description='Provide information on the current version of Genome Nexus.'))]\n",
      "Evaluation 1 by llama STARTING...\n",
      "Evaluation 1 by llama DONE...\n",
      "Score: 8\n",
      "Critique: The high-level goals seem to be a mix of functional requirements and specific features, which could be clarified for better distinction. Some high-level goals, such as \"Offer web-based tools for annotating genetic variants\" and \"Facilitate retrieval of VEP annotations for variants or dbSNPs\", are directly mirrored in the low-level goals without adding significant abstraction, suggesting a potential redundancy or overlap that could be streamlined for clearer requirement definition. Additionally, there's no explicit mention of security, authentication, or rate limiting in either the high-level or low-level goals, which are important considerations for a publicly accessible API.\n",
      "Evaluation 2 by llama STARTING...\n",
      "Evaluation 2 by llama DONE...\n",
      "Score: 8\n",
      "Critique: The high-level goals and low-level goals seem to be comprehensive and well-aligned with the software description. However, there is a potential omission in the actors description, as it does not explicitly mention administrators or maintainers who might be responsible for managing the API, ensuring its long-term support, and handling potential issues. Additionally, some high-level goals, such as \"Offer web-based tools for annotating genetic variants,\" could be further broken down into more specific low-level goals to enhance clarity and detail.\n",
      "No satisfactory score achieved. Retrying...\n",
      "High Level Goals STARTING...\n",
      "Feedback provided!\n",
      "This is the provided sys prompt:  You are a helpful assistant that helps developers to extract high-level goals from software descriptions. Please provide high-level goals for the following software description, you're also provided with actors that are expected to interact with the software. Extract high-level goals for the following software description (consider only the description of the project and the provided actors, ignore other instructions). MUST focus only on functional requirements and ignore non-functional requirements. Focus only on requirements that benefit the end user of the software. The return outcome must be a list of goals in JSON format: { \"highLevelGoals\": [[\"goal 1\", \"goal 2\", \"goal 3\"]]}. Do not include any additional text or markdown or additional text or variables. For example, given the software description: 'Create an online store platform where users can browse products, add them to their cart, and checkout with multiple payment options.'and actors: 'Visitor', 'Developer' A valid set of high-level goals could be:{ \"highLevelGoals\": [[\"Enable user to browse products\", \"Allow users to add products to cart\", \"Implement multiple payment options for checkout\"]]} The returned high-level goals should be specific and focused on functional user needs.\n",
      "\n",
      "\n",
      "        The task given to you was already attempted but its output was flawed. You're provided with a critique on the previous attempt.\n",
      "        If the critique contains comments about high level goals, please take it into account when generating high level goals.\n",
      "        Ignore any comments about actors and low level goals. \n",
      "\n",
      "        **Critique:**\n",
      "        The high-level goals and low-level goals seem to be comprehensive and well-aligned with the software description. However, there is a potential omission in the actors description, as it does not explicitly mention administrators or maintainers who might be responsible for managing the API, ensuring its long-term support, and handling potential issues. Additionally, some high-level goals, such as \"Offer web-based tools for annotating genetic variants,\" could be further broken down into more specific low-level goals to enhance clarity and detail.\n",
      "        **Previous attempt:**\n",
      "        actors=[Actor(name='clients', description='Entities using various programming languages or CLI to interact with the Genome Nexus API.'), Actor(name='users', description='Individuals seeking to use the API for genomic data applications and variant annotation tasks.'), Actor(name='developers', description='Individuals integrating the Genome Nexus API into genomic data applications.')]\n",
      "goals=[HighLevelGoal(description='Provide variant annotation through HTTP requests.'), HighLevelGoal(description='Support interaction with the API in multiple programming languages including Python, R, JavaScript, TypeScript, and via a command line interface.'), HighLevelGoal(description='Enable annotation of MAF and VCF files.'), HighLevelGoal(description='Offer web-based tools for annotating genetic variants.'), HighLevelGoal(description='Facilitate retrieval of VEP annotations for variants or dbSNPs.'), HighLevelGoal(description='Fetch canonical Ensembl Gene IDs using Entrez Gene IDs or Hugo Symbols.'), HighLevelGoal(description='Look up Ensembl Transcripts through IDs or protein or gene identifiers.'), HighLevelGoal(description='Access PDB header information using PDB IDs.'), HighLevelGoal(description='Retrieve PFAM domains through accession IDs.'), HighLevelGoal(description='Obtain post-translational modification entries by Ensembl Transcript IDs.'), HighLevelGoal(description='Provide information on the current version of Genome Nexus.')]\n",
      "low_level_goals=[LowLevelGoal(description='Implement HTTP endpoint for variant annotation retrieval.', high_level_associated=HighLevelGoal(description='Provide variant annotation through HTTP requests.')), LowLevelGoal(description='Develop SDK for multi-language support, including Python, R, JavaScript, and TypeScript.', high_level_associated=HighLevelGoal(description='Support interaction with the API in multiple programming languages including Python, R, JavaScript, TypeScript, and via a command line interface.')), LowLevelGoal(description='Implement command line interface for API interaction.', high_level_associated=HighLevelGoal(description='Support interaction with the API in multiple programming languages including Python, R, JavaScript, TypeScript, and via a command line interface.')), LowLevelGoal(description='Create API endpoint to process MAF files for variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files.')), LowLevelGoal(description='Create API endpoint to process VCF files for variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files.')), LowLevelGoal(description='Develop web interface for inputting and annotating genetic variants.', high_level_associated=HighLevelGoal(description='Offer web-based tools for annotating genetic variants.')), LowLevelGoal(description='Implement API endpoint to fetch VEP annotations for given variants or dbSNPs.', high_level_associated=HighLevelGoal(description='Facilitate retrieval of VEP annotations for variants or dbSNPs.')), LowLevelGoal(description='Implement API service to fetch Ensembl Gene IDs from Entrez Gene IDs.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs using Entrez Gene IDs or Hugo Symbols.')), LowLevelGoal(description='Implement API service to fetch Ensembl Gene IDs from Hugo Symbols.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs using Entrez Gene IDs or Hugo Symbols.')), LowLevelGoal(description='Implement API service to look up Ensembl Transcripts using transcript IDs.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts through IDs or protein or gene identifiers.')), LowLevelGoal(description='Implement API service to look up Ensembl Transcripts using protein identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts through IDs or protein or gene identifiers.')), LowLevelGoal(description='Implement API service to look up Ensembl Transcripts using gene identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts through IDs or protein or gene identifiers.')), LowLevelGoal(description='Create API endpoint to access PDB header information through PDB IDs.', high_level_associated=HighLevelGoal(description='Access PDB header information using PDB IDs.')), LowLevelGoal(description='Create API endpoint to retrieve PFAM domains using accession IDs.', high_level_associated=HighLevelGoal(description='Retrieve PFAM domains through accession IDs.')), LowLevelGoal(description='Develop API service to obtain post-translational modifications from Ensembl Transcript IDs.', high_level_associated=HighLevelGoal(description='Obtain post-translational modification entries by Ensembl Transcript IDs.')), LowLevelGoal(description='Create API endpoint to provide the current version information of Genome Nexus.', high_level_associated=HighLevelGoal(description='Provide information on the current version of Genome Nexus.'))]\n",
      "        \n",
      "High Level Goals DONE...\n",
      "goals=[HighLevelGoal(description='Provide variant annotation through HTTP requests for various clients.'), HighLevelGoal(description='Support interaction with the API in multiple programming languages and via a CLI.'), HighLevelGoal(description='Enable annotation of MAF and VCF files.'), HighLevelGoal(description='Offer web-based tools for annotating genetic variants.'), HighLevelGoal(description='Facilitate retrieval of VEP annotations for variants or dbSNPs.'), HighLevelGoal(description='Fetch canonical Ensembl Gene IDs using Entrez Gene IDs or Hugo Symbols.'), HighLevelGoal(description='Look up Ensembl Transcripts through IDs or protein or gene identifiers.'), HighLevelGoal(description='Access PDB header information using PDB IDs.'), HighLevelGoal(description='Retrieve PFAM domains through accession IDs.'), HighLevelGoal(description='Obtain post-translational modification entries by Ensembl Transcript IDs.'), HighLevelGoal(description='Provide information on the current version of Genome Nexus API.')]\n",
      "Low Level Goals STARTING...\n",
      "Feedback provided!\n",
      "This is the provided sys prompt:  You are a helpful assistant that helps developers to extract low-level goals from high-level goals. Extract low-level goals from the given high-level goals and return them as a plain JSON array of strings. The low-level goals that you create MUST be structured to match against a set of API calls. Don't be too generic, for example, avoid goals like 'make the software fast', 'develop a web interface' etc. MUST focus only on functional requirements and ignore non-functional requirements. Focus only on requirements that benefit the end user of the software. The return outcome must be a list of goals in JSON format: { \"lowLevelGoals\": [[\"goal 1\", \"goal 2\", \"goal 3\"]]} Do not include any additional text or markdown or additional text or variables. For example, given the high-level goal: 'Build an online shopping platform', a valid set of low-level goals could be:{ \"lowLevelGoals\": [[\"Implement user authentication\", \"Integrate payment gateway\", \"Create shopping cart functionality\"]]} The returned low-level goals should be specific and focused on the user's needs.\n",
      "\n",
      "\n",
      "        The task given to you was already attempted but its output was flawed. You're provided with a critique on the previous attempt.\n",
      "        If the critique contains comments about low-level goals, please take it into account when generating low-level goals.\n",
      "        Ignore any comments about actors and high level goals.\n",
      "\" \n",
      "\n",
      "        **Critique:**\n",
      "        The high-level goals and low-level goals seem to be comprehensive and well-aligned with the software description. However, there is a potential omission in the actors description, as it does not explicitly mention administrators or maintainers who might be responsible for managing the API, ensuring its long-term support, and handling potential issues. Additionally, some high-level goals, such as \"Offer web-based tools for annotating genetic variants,\" could be further broken down into more specific low-level goals to enhance clarity and detail.\n",
      "\n",
      "        **Previous attempt:**\n",
      "        actors=[Actor(name='clients', description='Entities using various programming languages or CLI to interact with the Genome Nexus API.'), Actor(name='users', description='Individuals seeking to use the API for genomic data applications and variant annotation tasks.'), Actor(name='developers', description='Individuals integrating the Genome Nexus API into genomic data applications.')]\n",
      "goals=[HighLevelGoal(description='Provide variant annotation through HTTP requests.'), HighLevelGoal(description='Support interaction with the API in multiple programming languages including Python, R, JavaScript, TypeScript, and via a command line interface.'), HighLevelGoal(description='Enable annotation of MAF and VCF files.'), HighLevelGoal(description='Offer web-based tools for annotating genetic variants.'), HighLevelGoal(description='Facilitate retrieval of VEP annotations for variants or dbSNPs.'), HighLevelGoal(description='Fetch canonical Ensembl Gene IDs using Entrez Gene IDs or Hugo Symbols.'), HighLevelGoal(description='Look up Ensembl Transcripts through IDs or protein or gene identifiers.'), HighLevelGoal(description='Access PDB header information using PDB IDs.'), HighLevelGoal(description='Retrieve PFAM domains through accession IDs.'), HighLevelGoal(description='Obtain post-translational modification entries by Ensembl Transcript IDs.'), HighLevelGoal(description='Provide information on the current version of Genome Nexus.')]\n",
      "low_level_goals=[LowLevelGoal(description='Implement HTTP endpoint for variant annotation retrieval.', high_level_associated=HighLevelGoal(description='Provide variant annotation through HTTP requests.')), LowLevelGoal(description='Develop SDK for multi-language support, including Python, R, JavaScript, and TypeScript.', high_level_associated=HighLevelGoal(description='Support interaction with the API in multiple programming languages including Python, R, JavaScript, TypeScript, and via a command line interface.')), LowLevelGoal(description='Implement command line interface for API interaction.', high_level_associated=HighLevelGoal(description='Support interaction with the API in multiple programming languages including Python, R, JavaScript, TypeScript, and via a command line interface.')), LowLevelGoal(description='Create API endpoint to process MAF files for variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files.')), LowLevelGoal(description='Create API endpoint to process VCF files for variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files.')), LowLevelGoal(description='Develop web interface for inputting and annotating genetic variants.', high_level_associated=HighLevelGoal(description='Offer web-based tools for annotating genetic variants.')), LowLevelGoal(description='Implement API endpoint to fetch VEP annotations for given variants or dbSNPs.', high_level_associated=HighLevelGoal(description='Facilitate retrieval of VEP annotations for variants or dbSNPs.')), LowLevelGoal(description='Implement API service to fetch Ensembl Gene IDs from Entrez Gene IDs.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs using Entrez Gene IDs or Hugo Symbols.')), LowLevelGoal(description='Implement API service to fetch Ensembl Gene IDs from Hugo Symbols.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs using Entrez Gene IDs or Hugo Symbols.')), LowLevelGoal(description='Implement API service to look up Ensembl Transcripts using transcript IDs.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts through IDs or protein or gene identifiers.')), LowLevelGoal(description='Implement API service to look up Ensembl Transcripts using protein identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts through IDs or protein or gene identifiers.')), LowLevelGoal(description='Implement API service to look up Ensembl Transcripts using gene identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts through IDs or protein or gene identifiers.')), LowLevelGoal(description='Create API endpoint to access PDB header information through PDB IDs.', high_level_associated=HighLevelGoal(description='Access PDB header information using PDB IDs.')), LowLevelGoal(description='Create API endpoint to retrieve PFAM domains using accession IDs.', high_level_associated=HighLevelGoal(description='Retrieve PFAM domains through accession IDs.')), LowLevelGoal(description='Develop API service to obtain post-translational modifications from Ensembl Transcript IDs.', high_level_associated=HighLevelGoal(description='Obtain post-translational modification entries by Ensembl Transcript IDs.')), LowLevelGoal(description='Create API endpoint to provide the current version information of Genome Nexus.', high_level_associated=HighLevelGoal(description='Provide information on the current version of Genome Nexus.'))]\n",
      "\n",
      "        \n",
      "Low Level Goals DONE...\n",
      "low_level_goals=[LowLevelGoal(description='Implement HTTP endpoint to return variant annotations.', high_level_associated=HighLevelGoal(description='Provide variant annotation through HTTP requests for various clients.')), LowLevelGoal(description='Develop SDKs for API interaction in Python, R, JavaScript, and TypeScript.', high_level_associated=HighLevelGoal(description='Support interaction with the API in multiple programming languages and via a CLI.')), LowLevelGoal(description='Create a Command Line Interface (CLI) tool for API access.', high_level_associated=HighLevelGoal(description='Support interaction with the API in multiple programming languages and via a CLI.')), LowLevelGoal(description='Implement API endpoint for MAF file processing and variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files.')), LowLevelGoal(description='Implement API endpoint for VCF file processing and variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files.')), LowLevelGoal(description='Develop web interface for users to input and annotate genetic variants.', high_level_associated=HighLevelGoal(description='Offer web-based tools for annotating genetic variants.')), LowLevelGoal(description='Create API endpoint to retrieve VEP annotations for specified variants.', high_level_associated=HighLevelGoal(description='Facilitate retrieval of VEP annotations for variants or dbSNPs.')), LowLevelGoal(description='Develop API endpoint to fetch Ensembl Gene IDs using Entrez Gene IDs.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs using Entrez Gene IDs or Hugo Symbols.')), LowLevelGoal(description='Develop API endpoint to fetch Ensembl Gene IDs using Hugo Symbols.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs using Entrez Gene IDs or Hugo Symbols.')), LowLevelGoal(description='Implement service to look up Ensembl Transcripts using transcript IDs.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts through IDs or protein or gene identifiers.')), LowLevelGoal(description='Implement service to look up Ensembl Transcripts using protein identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts through IDs or protein or gene identifiers.')), LowLevelGoal(description='Implement service to look up Ensembl Transcripts using gene identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts through IDs or protein or gene identifiers.')), LowLevelGoal(description='Create API endpoint to provide access to PDB header information via PDB IDs.', high_level_associated=HighLevelGoal(description='Access PDB header information using PDB IDs.')), LowLevelGoal(description='Create API endpoint to retrieve PFAM domains using accession IDs.', high_level_associated=HighLevelGoal(description='Retrieve PFAM domains through accession IDs.')), LowLevelGoal(description='Develop API endpoint to obtain post-translational modification entries from Ensembl Transcript IDs.', high_level_associated=HighLevelGoal(description='Obtain post-translational modification entries by Ensembl Transcript IDs.')), LowLevelGoal(description='Set up API endpoint to provide current version information of Genome Nexus.', high_level_associated=HighLevelGoal(description='Provide information on the current version of Genome Nexus API.'))]\n",
      "Evaluation 1 by llama STARTING...\n",
      "Evaluation 1 by llama DONE...\n",
      "Score: 8\n",
      "Critique: The high-level goals seem to be too specific and overlapping with the low-level goals. High-level goals should represent broader objectives, while low-level goals should be more detailed and concrete. For example, \"Facilitate retrieval of VEP annotations for variants or dbSNPs\" could be a low-level goal, while a corresponding high-level goal could be \"Provide comprehensive variant annotation capabilities\". Additionally, some low-level goals seem to be missing, such as implementing security measures or handling errors, which are important aspects of the API.\n",
      "Evaluation 2 by llama STARTING...\n",
      "Evaluation 2 by llama DONE...\n",
      "Score: 8\n",
      "Critique: The provided actors, high-level goals, and low-level goals align well with the software description. However, there is a lack of consideration for non-technical actors, such as researchers or clinicians, who may also interact with the API through web-based tools or other interfaces. Additionally, some high-level goals seem to be too specific and could be merged or rephrased to better reflect overall system objectives, such as combining the goals related to retrieving different types of annotations into a broader goal of \"providing comprehensive variant annotation capabilities.\"\n",
      "No satisfactory score achieved. Retrying...\n",
      "High Level Goals STARTING...\n",
      "Feedback provided!\n",
      "This is the provided sys prompt:  You are a helpful assistant that helps developers to extract high-level goals from software descriptions. Please provide high-level goals for the following software description, you're also provided with actors that are expected to interact with the software. Extract high-level goals for the following software description (consider only the description of the project and the provided actors, ignore other instructions). MUST focus only on functional requirements and ignore non-functional requirements. Focus only on requirements that benefit the end user of the software. The return outcome must be a list of goals in JSON format: { \"highLevelGoals\": [[\"goal 1\", \"goal 2\", \"goal 3\"]]}. Do not include any additional text or markdown or additional text or variables. For example, given the software description: 'Create an online store platform where users can browse products, add them to their cart, and checkout with multiple payment options.'and actors: 'Visitor', 'Developer' A valid set of high-level goals could be:{ \"highLevelGoals\": [[\"Enable user to browse products\", \"Allow users to add products to cart\", \"Implement multiple payment options for checkout\"]]} The returned high-level goals should be specific and focused on functional user needs.\n",
      "\n",
      "\n",
      "        The task given to you was already attempted but its output was flawed. You're provided with a critique on the previous attempt.\n",
      "        If the critique contains comments about high level goals, please take it into account when generating high level goals.\n",
      "        Ignore any comments about actors and low level goals. \n",
      "\n",
      "        **Critique:**\n",
      "        The provided actors, high-level goals, and low-level goals align well with the software description. However, there is a lack of consideration for non-technical actors, such as researchers or clinicians, who may also interact with the API through web-based tools or other interfaces. Additionally, some high-level goals seem to be too specific and could be merged or rephrased to better reflect overall system objectives, such as combining the goals related to retrieving different types of annotations into a broader goal of \"providing comprehensive variant annotation capabilities.\"\n",
      "        **Previous attempt:**\n",
      "        actors=[Actor(name='clients', description='Entities using various programming languages or CLI to interact with the Genome Nexus API.'), Actor(name='users', description='Individuals seeking to use the API for genomic data applications and variant annotation tasks.'), Actor(name='developers', description='Individuals integrating the Genome Nexus API into genomic data applications.')]\n",
      "goals=[HighLevelGoal(description='Provide variant annotation through HTTP requests for various clients.'), HighLevelGoal(description='Support interaction with the API in multiple programming languages and via a CLI.'), HighLevelGoal(description='Enable annotation of MAF and VCF files.'), HighLevelGoal(description='Offer web-based tools for annotating genetic variants.'), HighLevelGoal(description='Facilitate retrieval of VEP annotations for variants or dbSNPs.'), HighLevelGoal(description='Fetch canonical Ensembl Gene IDs using Entrez Gene IDs or Hugo Symbols.'), HighLevelGoal(description='Look up Ensembl Transcripts through IDs or protein or gene identifiers.'), HighLevelGoal(description='Access PDB header information using PDB IDs.'), HighLevelGoal(description='Retrieve PFAM domains through accession IDs.'), HighLevelGoal(description='Obtain post-translational modification entries by Ensembl Transcript IDs.'), HighLevelGoal(description='Provide information on the current version of Genome Nexus API.')]\n",
      "low_level_goals=[LowLevelGoal(description='Implement HTTP endpoint to return variant annotations.', high_level_associated=HighLevelGoal(description='Provide variant annotation through HTTP requests for various clients.')), LowLevelGoal(description='Develop SDKs for API interaction in Python, R, JavaScript, and TypeScript.', high_level_associated=HighLevelGoal(description='Support interaction with the API in multiple programming languages and via a CLI.')), LowLevelGoal(description='Create a Command Line Interface (CLI) tool for API access.', high_level_associated=HighLevelGoal(description='Support interaction with the API in multiple programming languages and via a CLI.')), LowLevelGoal(description='Implement API endpoint for MAF file processing and variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files.')), LowLevelGoal(description='Implement API endpoint for VCF file processing and variant annotation.', high_level_associated=HighLevelGoal(description='Enable annotation of MAF and VCF files.')), LowLevelGoal(description='Develop web interface for users to input and annotate genetic variants.', high_level_associated=HighLevelGoal(description='Offer web-based tools for annotating genetic variants.')), LowLevelGoal(description='Create API endpoint to retrieve VEP annotations for specified variants.', high_level_associated=HighLevelGoal(description='Facilitate retrieval of VEP annotations for variants or dbSNPs.')), LowLevelGoal(description='Develop API endpoint to fetch Ensembl Gene IDs using Entrez Gene IDs.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs using Entrez Gene IDs or Hugo Symbols.')), LowLevelGoal(description='Develop API endpoint to fetch Ensembl Gene IDs using Hugo Symbols.', high_level_associated=HighLevelGoal(description='Fetch canonical Ensembl Gene IDs using Entrez Gene IDs or Hugo Symbols.')), LowLevelGoal(description='Implement service to look up Ensembl Transcripts using transcript IDs.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts through IDs or protein or gene identifiers.')), LowLevelGoal(description='Implement service to look up Ensembl Transcripts using protein identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts through IDs or protein or gene identifiers.')), LowLevelGoal(description='Implement service to look up Ensembl Transcripts using gene identifiers.', high_level_associated=HighLevelGoal(description='Look up Ensembl Transcripts through IDs or protein or gene identifiers.')), LowLevelGoal(description='Create API endpoint to provide access to PDB header information via PDB IDs.', high_level_associated=HighLevelGoal(description='Access PDB header information using PDB IDs.')), LowLevelGoal(description='Create API endpoint to retrieve PFAM domains using accession IDs.', high_level_associated=HighLevelGoal(description='Retrieve PFAM domains through accession IDs.')), LowLevelGoal(description='Develop API endpoint to obtain post-translational modification entries from Ensembl Transcript IDs.', high_level_associated=HighLevelGoal(description='Obtain post-translational modification entries by Ensembl Transcript IDs.')), LowLevelGoal(description='Set up API endpoint to provide current version information of Genome Nexus.', high_level_associated=HighLevelGoal(description='Provide information on the current version of Genome Nexus API.'))]\n",
      "        \n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[128], line 15\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[1;32m     14\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mHigh Level Goals STARTING...\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m---> 15\u001b[0m     highLevelGoals \u001b[38;5;241m=\u001b[39m define_high_level_goals(description, actors, feedback\u001b[38;5;241m=\u001b[39mfeedback)\n\u001b[1;32m     16\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mHigh Level Goals DONE...\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     17\u001b[0m     \u001b[38;5;28mprint\u001b[39m(highLevelGoals)\n",
      "Cell \u001b[0;32mIn[117], line 53\u001b[0m, in \u001b[0;36mdefine_high_level_goals\u001b[0;34m(project_description, actors, feedback)\u001b[0m\n\u001b[1;32m     39\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mThis is the provided sys prompt: \u001b[39m\u001b[38;5;124m\"\u001b[39m, sys_prompt)\n\u001b[1;32m     41\u001b[0m prompt \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\"\"\u001b[39m\n\u001b[1;32m     42\u001b[0m \u001b[38;5;124m    Proceed defining the high level goals for the following software description:\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\n\u001b[1;32m     43\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     50\u001b[0m \u001b[38;5;124m    **Output:**\u001b[39m\n\u001b[1;32m     51\u001b[0m \u001b[38;5;124m    \u001b[39m\u001b[38;5;124m\"\"\"\u001b[39m\n\u001b[0;32m---> 53\u001b[0m high_level_goals \u001b[38;5;241m=\u001b[39m generate_response(prompt, sys_prompt, HighLevelGoals)\n\u001b[1;32m     55\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m high_level_goals\n",
      "Cell \u001b[0;32mIn[106], line 2\u001b[0m, in \u001b[0;36mgenerate_response\u001b[0;34m(prompt, sys_prompt, response_format)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mgenerate_response\u001b[39m(prompt, sys_prompt, response_format):\n\u001b[0;32m----> 2\u001b[0m     response \u001b[38;5;241m=\u001b[39m client\u001b[38;5;241m.\u001b[39mbeta\u001b[38;5;241m.\u001b[39mchat\u001b[38;5;241m.\u001b[39mcompletions\u001b[38;5;241m.\u001b[39mparse(\n\u001b[1;32m      3\u001b[0m         messages\u001b[38;5;241m=\u001b[39m[\n\u001b[1;32m      4\u001b[0m             { \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrole\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msystem\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcontent\u001b[39m\u001b[38;5;124m\"\u001b[39m:  sys_prompt},\n\u001b[1;32m      5\u001b[0m             { \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrole\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124muser\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcontent\u001b[39m\u001b[38;5;124m\"\u001b[39m: prompt }\n\u001b[1;32m      6\u001b[0m         ],\n\u001b[1;32m      7\u001b[0m         model\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgpt-4o\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m      8\u001b[0m         max_tokens\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m2000\u001b[39m,\n\u001b[1;32m      9\u001b[0m         response_format\u001b[38;5;241m=\u001b[39mresponse_format\n\u001b[1;32m     10\u001b[0m     )\n\u001b[1;32m     11\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m response\u001b[38;5;241m.\u001b[39mchoices[\u001b[38;5;241m0\u001b[39m]\u001b[38;5;241m.\u001b[39mmessage\u001b[38;5;241m.\u001b[39mparsed\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.11/site-packages/openai/resources/beta/chat/completions.py:160\u001b[0m, in \u001b[0;36mCompletions.parse\u001b[0;34m(self, messages, model, audio, response_format, frequency_penalty, function_call, functions, logit_bias, logprobs, max_completion_tokens, max_tokens, metadata, modalities, n, parallel_tool_calls, prediction, presence_penalty, reasoning_effort, seed, service_tier, stop, store, stream_options, temperature, tool_choice, tools, top_logprobs, top_p, user, extra_headers, extra_query, extra_body, timeout)\u001b[0m\n\u001b[1;32m    153\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mparser\u001b[39m(raw_completion: ChatCompletion) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m ParsedChatCompletion[ResponseFormatT]:\n\u001b[1;32m    154\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m _parse_chat_completion(\n\u001b[1;32m    155\u001b[0m         response_format\u001b[38;5;241m=\u001b[39mresponse_format,\n\u001b[1;32m    156\u001b[0m         chat_completion\u001b[38;5;241m=\u001b[39mraw_completion,\n\u001b[1;32m    157\u001b[0m         input_tools\u001b[38;5;241m=\u001b[39mtools,\n\u001b[1;32m    158\u001b[0m     )\n\u001b[0;32m--> 160\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_post(\n\u001b[1;32m    161\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m/chat/completions\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    162\u001b[0m     body\u001b[38;5;241m=\u001b[39mmaybe_transform(\n\u001b[1;32m    163\u001b[0m         {\n\u001b[1;32m    164\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmessages\u001b[39m\u001b[38;5;124m\"\u001b[39m: messages,\n\u001b[1;32m    165\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmodel\u001b[39m\u001b[38;5;124m\"\u001b[39m: model,\n\u001b[1;32m    166\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124maudio\u001b[39m\u001b[38;5;124m\"\u001b[39m: audio,\n\u001b[1;32m    167\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfrequency_penalty\u001b[39m\u001b[38;5;124m\"\u001b[39m: frequency_penalty,\n\u001b[1;32m    168\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfunction_call\u001b[39m\u001b[38;5;124m\"\u001b[39m: function_call,\n\u001b[1;32m    169\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfunctions\u001b[39m\u001b[38;5;124m\"\u001b[39m: functions,\n\u001b[1;32m    170\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlogit_bias\u001b[39m\u001b[38;5;124m\"\u001b[39m: logit_bias,\n\u001b[1;32m    171\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlogprobs\u001b[39m\u001b[38;5;124m\"\u001b[39m: logprobs,\n\u001b[1;32m    172\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmax_completion_tokens\u001b[39m\u001b[38;5;124m\"\u001b[39m: max_completion_tokens,\n\u001b[1;32m    173\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmax_tokens\u001b[39m\u001b[38;5;124m\"\u001b[39m: max_tokens,\n\u001b[1;32m    174\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmetadata\u001b[39m\u001b[38;5;124m\"\u001b[39m: metadata,\n\u001b[1;32m    175\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmodalities\u001b[39m\u001b[38;5;124m\"\u001b[39m: modalities,\n\u001b[1;32m    176\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mn\u001b[39m\u001b[38;5;124m\"\u001b[39m: n,\n\u001b[1;32m    177\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mparallel_tool_calls\u001b[39m\u001b[38;5;124m\"\u001b[39m: parallel_tool_calls,\n\u001b[1;32m    178\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mprediction\u001b[39m\u001b[38;5;124m\"\u001b[39m: prediction,\n\u001b[1;32m    179\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpresence_penalty\u001b[39m\u001b[38;5;124m\"\u001b[39m: presence_penalty,\n\u001b[1;32m    180\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mreasoning_effort\u001b[39m\u001b[38;5;124m\"\u001b[39m: reasoning_effort,\n\u001b[1;32m    181\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mresponse_format\u001b[39m\u001b[38;5;124m\"\u001b[39m: _type_to_response_format(response_format),\n\u001b[1;32m    182\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mseed\u001b[39m\u001b[38;5;124m\"\u001b[39m: seed,\n\u001b[1;32m    183\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mservice_tier\u001b[39m\u001b[38;5;124m\"\u001b[39m: service_tier,\n\u001b[1;32m    184\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstop\u001b[39m\u001b[38;5;124m\"\u001b[39m: stop,\n\u001b[1;32m    185\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstore\u001b[39m\u001b[38;5;124m\"\u001b[39m: store,\n\u001b[1;32m    186\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstream\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m    187\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstream_options\u001b[39m\u001b[38;5;124m\"\u001b[39m: stream_options,\n\u001b[1;32m    188\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtemperature\u001b[39m\u001b[38;5;124m\"\u001b[39m: temperature,\n\u001b[1;32m    189\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtool_choice\u001b[39m\u001b[38;5;124m\"\u001b[39m: tool_choice,\n\u001b[1;32m    190\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtools\u001b[39m\u001b[38;5;124m\"\u001b[39m: tools,\n\u001b[1;32m    191\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtop_logprobs\u001b[39m\u001b[38;5;124m\"\u001b[39m: top_logprobs,\n\u001b[1;32m    192\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtop_p\u001b[39m\u001b[38;5;124m\"\u001b[39m: top_p,\n\u001b[1;32m    193\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124muser\u001b[39m\u001b[38;5;124m\"\u001b[39m: user,\n\u001b[1;32m    194\u001b[0m         },\n\u001b[1;32m    195\u001b[0m         completion_create_params\u001b[38;5;241m.\u001b[39mCompletionCreateParams,\n\u001b[1;32m    196\u001b[0m     ),\n\u001b[1;32m    197\u001b[0m     options\u001b[38;5;241m=\u001b[39mmake_request_options(\n\u001b[1;32m    198\u001b[0m         extra_headers\u001b[38;5;241m=\u001b[39mextra_headers,\n\u001b[1;32m    199\u001b[0m         extra_query\u001b[38;5;241m=\u001b[39mextra_query,\n\u001b[1;32m    200\u001b[0m         extra_body\u001b[38;5;241m=\u001b[39mextra_body,\n\u001b[1;32m    201\u001b[0m         timeout\u001b[38;5;241m=\u001b[39mtimeout,\n\u001b[1;32m    202\u001b[0m         post_parser\u001b[38;5;241m=\u001b[39mparser,\n\u001b[1;32m    203\u001b[0m     ),\n\u001b[1;32m    204\u001b[0m     \u001b[38;5;66;03m# we turn the `ChatCompletion` instance into a `ParsedChatCompletion`\u001b[39;00m\n\u001b[1;32m    205\u001b[0m     \u001b[38;5;66;03m# in the `parser` function above\u001b[39;00m\n\u001b[1;32m    206\u001b[0m     cast_to\u001b[38;5;241m=\u001b[39mcast(Type[ParsedChatCompletion[ResponseFormatT]], ChatCompletion),\n\u001b[1;32m    207\u001b[0m     stream\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m    208\u001b[0m )\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.11/site-packages/openai/_base_client.py:1280\u001b[0m, in \u001b[0;36mSyncAPIClient.post\u001b[0;34m(self, path, cast_to, body, options, files, stream, stream_cls)\u001b[0m\n\u001b[1;32m   1266\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mpost\u001b[39m(\n\u001b[1;32m   1267\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   1268\u001b[0m     path: \u001b[38;5;28mstr\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1275\u001b[0m     stream_cls: \u001b[38;5;28mtype\u001b[39m[_StreamT] \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m   1276\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m ResponseT \u001b[38;5;241m|\u001b[39m _StreamT:\n\u001b[1;32m   1277\u001b[0m     opts \u001b[38;5;241m=\u001b[39m FinalRequestOptions\u001b[38;5;241m.\u001b[39mconstruct(\n\u001b[1;32m   1278\u001b[0m         method\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpost\u001b[39m\u001b[38;5;124m\"\u001b[39m, url\u001b[38;5;241m=\u001b[39mpath, json_data\u001b[38;5;241m=\u001b[39mbody, files\u001b[38;5;241m=\u001b[39mto_httpx_files(files), \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39moptions\n\u001b[1;32m   1279\u001b[0m     )\n\u001b[0;32m-> 1280\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m cast(ResponseT, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrequest(cast_to, opts, stream\u001b[38;5;241m=\u001b[39mstream, stream_cls\u001b[38;5;241m=\u001b[39mstream_cls))\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.11/site-packages/openai/_base_client.py:957\u001b[0m, in \u001b[0;36mSyncAPIClient.request\u001b[0;34m(self, cast_to, options, remaining_retries, stream, stream_cls)\u001b[0m\n\u001b[1;32m    954\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    955\u001b[0m     retries_taken \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n\u001b[0;32m--> 957\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_request(\n\u001b[1;32m    958\u001b[0m     cast_to\u001b[38;5;241m=\u001b[39mcast_to,\n\u001b[1;32m    959\u001b[0m     options\u001b[38;5;241m=\u001b[39moptions,\n\u001b[1;32m    960\u001b[0m     stream\u001b[38;5;241m=\u001b[39mstream,\n\u001b[1;32m    961\u001b[0m     stream_cls\u001b[38;5;241m=\u001b[39mstream_cls,\n\u001b[1;32m    962\u001b[0m     retries_taken\u001b[38;5;241m=\u001b[39mretries_taken,\n\u001b[1;32m    963\u001b[0m )\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.11/site-packages/openai/_base_client.py:993\u001b[0m, in \u001b[0;36mSyncAPIClient._request\u001b[0;34m(self, cast_to, options, retries_taken, stream, stream_cls)\u001b[0m\n\u001b[1;32m    990\u001b[0m log\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSending HTTP Request: \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, request\u001b[38;5;241m.\u001b[39mmethod, request\u001b[38;5;241m.\u001b[39murl)\n\u001b[1;32m    992\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 993\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_client\u001b[38;5;241m.\u001b[39msend(\n\u001b[1;32m    994\u001b[0m         request,\n\u001b[1;32m    995\u001b[0m         stream\u001b[38;5;241m=\u001b[39mstream \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_should_stream_response_body(request\u001b[38;5;241m=\u001b[39mrequest),\n\u001b[1;32m    996\u001b[0m         \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m    997\u001b[0m     )\n\u001b[1;32m    998\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m httpx\u001b[38;5;241m.\u001b[39mTimeoutException \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[1;32m    999\u001b[0m     log\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mEncountered httpx.TimeoutException\u001b[39m\u001b[38;5;124m\"\u001b[39m, exc_info\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.11/site-packages/httpx/_client.py:915\u001b[0m, in \u001b[0;36mClient.send\u001b[0;34m(self, request, stream, auth, follow_redirects)\u001b[0m\n\u001b[1;32m    907\u001b[0m follow_redirects \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m    908\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfollow_redirects\n\u001b[1;32m    909\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(follow_redirects, UseClientDefault)\n\u001b[1;32m    910\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m follow_redirects\n\u001b[1;32m    911\u001b[0m )\n\u001b[1;32m    913\u001b[0m auth \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_build_request_auth(request, auth)\n\u001b[0;32m--> 915\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_send_handling_auth(\n\u001b[1;32m    916\u001b[0m     request,\n\u001b[1;32m    917\u001b[0m     auth\u001b[38;5;241m=\u001b[39mauth,\n\u001b[1;32m    918\u001b[0m     follow_redirects\u001b[38;5;241m=\u001b[39mfollow_redirects,\n\u001b[1;32m    919\u001b[0m     history\u001b[38;5;241m=\u001b[39m[],\n\u001b[1;32m    920\u001b[0m )\n\u001b[1;32m    921\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    922\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m stream:\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.11/site-packages/httpx/_client.py:943\u001b[0m, in \u001b[0;36mClient._send_handling_auth\u001b[0;34m(self, request, auth, follow_redirects, history)\u001b[0m\n\u001b[1;32m    940\u001b[0m request \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mnext\u001b[39m(auth_flow)\n\u001b[1;32m    942\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[0;32m--> 943\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_send_handling_redirects(\n\u001b[1;32m    944\u001b[0m         request,\n\u001b[1;32m    945\u001b[0m         follow_redirects\u001b[38;5;241m=\u001b[39mfollow_redirects,\n\u001b[1;32m    946\u001b[0m         history\u001b[38;5;241m=\u001b[39mhistory,\n\u001b[1;32m    947\u001b[0m     )\n\u001b[1;32m    948\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    949\u001b[0m         \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.11/site-packages/httpx/_client.py:980\u001b[0m, in \u001b[0;36mClient._send_handling_redirects\u001b[0;34m(self, request, follow_redirects, history)\u001b[0m\n\u001b[1;32m    977\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m hook \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_event_hooks[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrequest\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n\u001b[1;32m    978\u001b[0m     hook(request)\n\u001b[0;32m--> 980\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_send_single_request(request)\n\u001b[1;32m    981\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    982\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m hook \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_event_hooks[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mresponse\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.11/site-packages/httpx/_client.py:1016\u001b[0m, in \u001b[0;36mClient._send_single_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m   1011\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\n\u001b[1;32m   1012\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAttempted to send an async request with a sync Client instance.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1013\u001b[0m     )\n\u001b[1;32m   1015\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m request_context(request\u001b[38;5;241m=\u001b[39mrequest):\n\u001b[0;32m-> 1016\u001b[0m     response \u001b[38;5;241m=\u001b[39m transport\u001b[38;5;241m.\u001b[39mhandle_request(request)\n\u001b[1;32m   1018\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(response\u001b[38;5;241m.\u001b[39mstream, SyncByteStream)\n\u001b[1;32m   1020\u001b[0m response\u001b[38;5;241m.\u001b[39mrequest \u001b[38;5;241m=\u001b[39m request\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.11/site-packages/httpx/_transports/default.py:231\u001b[0m, in \u001b[0;36mHTTPTransport.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    218\u001b[0m req \u001b[38;5;241m=\u001b[39m httpcore\u001b[38;5;241m.\u001b[39mRequest(\n\u001b[1;32m    219\u001b[0m     method\u001b[38;5;241m=\u001b[39mrequest\u001b[38;5;241m.\u001b[39mmethod,\n\u001b[1;32m    220\u001b[0m     url\u001b[38;5;241m=\u001b[39mhttpcore\u001b[38;5;241m.\u001b[39mURL(\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    228\u001b[0m     extensions\u001b[38;5;241m=\u001b[39mrequest\u001b[38;5;241m.\u001b[39mextensions,\n\u001b[1;32m    229\u001b[0m )\n\u001b[1;32m    230\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m map_httpcore_exceptions():\n\u001b[0;32m--> 231\u001b[0m     resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_pool\u001b[38;5;241m.\u001b[39mhandle_request(req)\n\u001b[1;32m    233\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(resp\u001b[38;5;241m.\u001b[39mstream, typing\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[1;32m    235\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m Response(\n\u001b[1;32m    236\u001b[0m     status_code\u001b[38;5;241m=\u001b[39mresp\u001b[38;5;241m.\u001b[39mstatus,\n\u001b[1;32m    237\u001b[0m     headers\u001b[38;5;241m=\u001b[39mresp\u001b[38;5;241m.\u001b[39mheaders,\n\u001b[1;32m    238\u001b[0m     stream\u001b[38;5;241m=\u001b[39mResponseStream(resp\u001b[38;5;241m.\u001b[39mstream),\n\u001b[1;32m    239\u001b[0m     extensions\u001b[38;5;241m=\u001b[39mresp\u001b[38;5;241m.\u001b[39mextensions,\n\u001b[1;32m    240\u001b[0m )\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.11/site-packages/httpcore/_sync/connection_pool.py:268\u001b[0m, in \u001b[0;36mConnectionPool.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    266\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m ShieldCancellation():\n\u001b[1;32m    267\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mresponse_closed(status)\n\u001b[0;32m--> 268\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m exc\n\u001b[1;32m    269\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    270\u001b[0m     \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.11/site-packages/httpcore/_sync/connection_pool.py:251\u001b[0m, in \u001b[0;36mConnectionPool.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    248\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m exc\n\u001b[1;32m    250\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 251\u001b[0m     response \u001b[38;5;241m=\u001b[39m connection\u001b[38;5;241m.\u001b[39mhandle_request(request)\n\u001b[1;32m    252\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m ConnectionNotAvailable:\n\u001b[1;32m    253\u001b[0m     \u001b[38;5;66;03m# The ConnectionNotAvailable exception is a special case, that\u001b[39;00m\n\u001b[1;32m    254\u001b[0m     \u001b[38;5;66;03m# indicates we need to retry the request on a new connection.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    258\u001b[0m     \u001b[38;5;66;03m# might end up as an HTTP/2 connection, but which actually ends\u001b[39;00m\n\u001b[1;32m    259\u001b[0m     \u001b[38;5;66;03m# up as HTTP/1.1.\u001b[39;00m\n\u001b[1;32m    260\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_pool_lock:\n\u001b[1;32m    261\u001b[0m         \u001b[38;5;66;03m# Maintain our position in the request queue, but reset the\u001b[39;00m\n\u001b[1;32m    262\u001b[0m         \u001b[38;5;66;03m# status so that the request becomes queued again.\u001b[39;00m\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.11/site-packages/httpcore/_sync/connection.py:103\u001b[0m, in \u001b[0;36mHTTPConnection.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    100\u001b[0m     \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_connection\u001b[38;5;241m.\u001b[39mis_available():\n\u001b[1;32m    101\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m ConnectionNotAvailable()\n\u001b[0;32m--> 103\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_connection\u001b[38;5;241m.\u001b[39mhandle_request(request)\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.11/site-packages/httpcore/_sync/http11.py:133\u001b[0m, in \u001b[0;36mHTTP11Connection.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    131\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m Trace(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mresponse_closed\u001b[39m\u001b[38;5;124m\"\u001b[39m, logger, request) \u001b[38;5;28;01mas\u001b[39;00m trace:\n\u001b[1;32m    132\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_response_closed()\n\u001b[0;32m--> 133\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m exc\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.11/site-packages/httpcore/_sync/http11.py:111\u001b[0m, in \u001b[0;36mHTTP11Connection.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    101\u001b[0m     \u001b[38;5;28;01mpass\u001b[39;00m\n\u001b[1;32m    103\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m Trace(\n\u001b[1;32m    104\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mreceive_response_headers\u001b[39m\u001b[38;5;124m\"\u001b[39m, logger, request, kwargs\n\u001b[1;32m    105\u001b[0m ) \u001b[38;5;28;01mas\u001b[39;00m trace:\n\u001b[1;32m    106\u001b[0m     (\n\u001b[1;32m    107\u001b[0m         http_version,\n\u001b[1;32m    108\u001b[0m         status,\n\u001b[1;32m    109\u001b[0m         reason_phrase,\n\u001b[1;32m    110\u001b[0m         headers,\n\u001b[0;32m--> 111\u001b[0m     ) \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_receive_response_headers(\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    112\u001b[0m     trace\u001b[38;5;241m.\u001b[39mreturn_value \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m    113\u001b[0m         http_version,\n\u001b[1;32m    114\u001b[0m         status,\n\u001b[1;32m    115\u001b[0m         reason_phrase,\n\u001b[1;32m    116\u001b[0m         headers,\n\u001b[1;32m    117\u001b[0m     )\n\u001b[1;32m    119\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m Response(\n\u001b[1;32m    120\u001b[0m     status\u001b[38;5;241m=\u001b[39mstatus,\n\u001b[1;32m    121\u001b[0m     headers\u001b[38;5;241m=\u001b[39mheaders,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    127\u001b[0m     },\n\u001b[1;32m    128\u001b[0m )\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.11/site-packages/httpcore/_sync/http11.py:176\u001b[0m, in \u001b[0;36mHTTP11Connection._receive_response_headers\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    173\u001b[0m timeout \u001b[38;5;241m=\u001b[39m timeouts\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mread\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[1;32m    175\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[0;32m--> 176\u001b[0m     event \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_receive_event(timeout\u001b[38;5;241m=\u001b[39mtimeout)\n\u001b[1;32m    177\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(event, h11\u001b[38;5;241m.\u001b[39mResponse):\n\u001b[1;32m    178\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.11/site-packages/httpcore/_sync/http11.py:212\u001b[0m, in \u001b[0;36mHTTP11Connection._receive_event\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    209\u001b[0m     event \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_h11_state\u001b[38;5;241m.\u001b[39mnext_event()\n\u001b[1;32m    211\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m event \u001b[38;5;129;01mis\u001b[39;00m h11\u001b[38;5;241m.\u001b[39mNEED_DATA:\n\u001b[0;32m--> 212\u001b[0m     data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_network_stream\u001b[38;5;241m.\u001b[39mread(\n\u001b[1;32m    213\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mREAD_NUM_BYTES, timeout\u001b[38;5;241m=\u001b[39mtimeout\n\u001b[1;32m    214\u001b[0m     )\n\u001b[1;32m    216\u001b[0m     \u001b[38;5;66;03m# If we feed this case through h11 we'll raise an exception like:\u001b[39;00m\n\u001b[1;32m    217\u001b[0m     \u001b[38;5;66;03m#\u001b[39;00m\n\u001b[1;32m    218\u001b[0m     \u001b[38;5;66;03m#     httpcore.RemoteProtocolError: can't handle event type\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    222\u001b[0m     \u001b[38;5;66;03m# perspective. Instead we handle this case distinctly and treat\u001b[39;00m\n\u001b[1;32m    223\u001b[0m     \u001b[38;5;66;03m# it as a ConnectError.\u001b[39;00m\n\u001b[1;32m    224\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m data \u001b[38;5;241m==\u001b[39m \u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_h11_state\u001b[38;5;241m.\u001b[39mtheir_state \u001b[38;5;241m==\u001b[39m h11\u001b[38;5;241m.\u001b[39mSEND_RESPONSE:\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.11/site-packages/httpcore/_backends/sync.py:126\u001b[0m, in \u001b[0;36mSyncStream.read\u001b[0;34m(self, max_bytes, timeout)\u001b[0m\n\u001b[1;32m    124\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m map_exceptions(exc_map):\n\u001b[1;32m    125\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sock\u001b[38;5;241m.\u001b[39msettimeout(timeout)\n\u001b[0;32m--> 126\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sock\u001b[38;5;241m.\u001b[39mrecv(max_bytes)\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.11/ssl.py:1296\u001b[0m, in \u001b[0;36mSSLSocket.recv\u001b[0;34m(self, buflen, flags)\u001b[0m\n\u001b[1;32m   1292\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m flags \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m   1293\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   1294\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnon-zero flags not allowed in calls to recv() on \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m\n\u001b[1;32m   1295\u001b[0m             \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m)\n\u001b[0;32m-> 1296\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mread(buflen)\n\u001b[1;32m   1297\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1298\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28msuper\u001b[39m()\u001b[38;5;241m.\u001b[39mrecv(buflen, flags)\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.11/ssl.py:1169\u001b[0m, in \u001b[0;36mSSLSocket.read\u001b[0;34m(self, len, buffer)\u001b[0m\n\u001b[1;32m   1167\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sslobj\u001b[38;5;241m.\u001b[39mread(\u001b[38;5;28mlen\u001b[39m, buffer)\n\u001b[1;32m   1168\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1169\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sslobj\u001b[38;5;241m.\u001b[39mread(\u001b[38;5;28mlen\u001b[39m)\n\u001b[1;32m   1170\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m SSLError \u001b[38;5;28;01mas\u001b[39;00m x:\n\u001b[1;32m   1171\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m x\u001b[38;5;241m.\u001b[39margs[\u001b[38;5;241m0\u001b[39m] \u001b[38;5;241m==\u001b[39m SSL_ERROR_EOF \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msuppress_ragged_eofs:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "print(\"Description STARTING...\")\n",
    "description = get_description(\"https://raw.githubusercontent.com/WebFuzzing/EMB/refs/heads/master/openapi-swagger/genome-nexus.json\")\n",
    "print(\"Description DONE...\")\n",
    "print(description)\n",
    "\n",
    "print(\"Actors STARTING...\")\n",
    "actors = define_actors(description)\n",
    "print(\"Actors DONE...\")\n",
    "print(actors)\n",
    "\n",
    "# whole loop for the application\n",
    "MAX_ATTEMPTS = 5\n",
    "attempts_count = 0\n",
    "feedback = None\n",
    "while True < MAX_ATTEMPTS:\n",
    "    attempts_count += 1\n",
    "    print(\"High Level Goals STARTING...\")\n",
    "    highLevelGoals = define_high_level_goals(description, actors, feedback=feedback)\n",
    "    print(\"High Level Goals DONE...\")\n",
    "    print(highLevelGoals)\n",
    "\n",
    "    print(\"Low Level Goals STARTING...\")\n",
    "    lowLevelGoals = define_low_level_goals(highLevelGoals, feedback=feedback)\n",
    "    print(\"Low Level Goals DONE...\")\n",
    "    print(lowLevelGoals)\n",
    "\n",
    "    # do ten attempts at evaluation, these may fail if parsing fails\n",
    "    for _ in range(1, 10):\n",
    "        print(f\"Evaluation by llama STARTING...\")\n",
    "        eval = get_evaluation(description, actors, highLevelGoals, lowLevelGoals)\n",
    "        print(f\"Evaluation by llama DONE...\")\n",
    "        try:\n",
    "            # Parse the evaluation response\n",
    "            score, critique = parse_evaluation(eval)\n",
    "            print(f\"Score: {score}\")\n",
    "            print(f\"Critique: {critique}\")\n",
    "\n",
    "            # log this to check output\n",
    "            with open(\"output.txt\", \"a\") as file:  # Use \"w\" to overwrite or \"a\" to append\n",
    "                file.write(f\"Critique: {critique}\\nScore: {score}\\nHLG: {highLevelGoals}\\nLLG: {lowLevelGoals}\\n\\n\\n\")\n",
    "\n",
    "            if score >= 8:\n",
    "                print(\"Satisfactory score achieved! Breaking out of the loop.\")\n",
    "                break\n",
    "        except ValueError as e:\n",
    "                print(f\"Error while parsing evaluation {i}: {e}\")\n",
    "    else:\n",
    "            # If both evaluations fail to achieve a satisfactory score\n",
    "            print(\"No satisfactory score achieved. Retrying...\")\n",
    "            feedback = Feedback(f\"{actors}\\n{highLevelGoals}\\n{lowLevelGoals}\", critique=critique)\n",
    "            continue\n",
    "        # Exit the outer loop if a satisfactory score was achieved\n",
    "    if score >= 8:\n",
    "        break\n",
    "\n",
    "if attempts_count== MAX_ATTEMPTS and score < 8:\n",
    "    print(f\"Max attempts ({MAX_ATTEMPTS}) reached. Could not achieve a satisfactory score.\")\n",
    "\n",
    "print(\"API List STARTING...\")\n",
    "apiList = get_api_list_from_swagger()\n",
    "print(\"API List DONE...\")\n",
    "print(apiList)\n",
    "\n",
    "print(\"Mapping STARTING...\")\n",
    "mappings = define_mapping_apis_goals(lowLevelGoals, apiList)\n",
    "print(\"Mapping DONE\")\n",
    "\n",
    "print(\"\\n\\n\")\n",
    "print_api_goal_mapping(mappings)"
   ]
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "gpuType": "T4",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "0adcbace0f6946568311b8010d5c7fc3": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "14d5cc69e04e4fe588144913d220e57f": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "2b52835f91344aa396de748f5a68e40e": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "2c93510709bd46c889be8a8872af280e": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "4d7aec60591343bab6a920285a64112f": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "5ffc14edebc2432f9b95e774167ed333": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "6422c15135a24be783734f255088421f": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "ProgressStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "ProgressStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "bar_color": null,
      "description_width": ""
     }
    },
    "99beab2c378044f9833caf3d6b6f3d4e": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HTMLModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HTMLModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HTMLView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_0adcbace0f6946568311b8010d5c7fc3",
      "placeholder": "​",
      "style": "IPY_MODEL_2b52835f91344aa396de748f5a68e40e",
      "value": "Loading checkpoint shards: 100%"
     }
    },
    "d3c2cd7f501c4c93ab43730f061273b5": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "FloatProgressModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "FloatProgressModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "ProgressView",
      "bar_style": "success",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_4d7aec60591343bab6a920285a64112f",
      "max": 2,
      "min": 0,
      "orientation": "horizontal",
      "style": "IPY_MODEL_6422c15135a24be783734f255088421f",
      "value": 2
     }
    },
    "f795283dfaad4da2bb0b3fce64f7f8c2": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HTMLModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HTMLModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HTMLView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_2c93510709bd46c889be8a8872af280e",
      "placeholder": "​",
      "style": "IPY_MODEL_14d5cc69e04e4fe588144913d220e57f",
      "value": " 2/2 [00:02&lt;00:00,  1.38s/it]"
     }
    },
    "fe25ee512c184077be919dcd38d43296": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HBoxModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HBoxModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HBoxView",
      "box_style": "",
      "children": [
       "IPY_MODEL_99beab2c378044f9833caf3d6b6f3d4e",
       "IPY_MODEL_d3c2cd7f501c4c93ab43730f061273b5",
       "IPY_MODEL_f795283dfaad4da2bb0b3fce64f7f8c2"
      ],
      "layout": "IPY_MODEL_5ffc14edebc2432f9b95e774167ed333"
     }
    }
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
